<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Analysis Using R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />

<link rel="next" href="lecture-slides.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2024 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#download-example-data"><i class="fa fa-check"></i>Download example data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#bonus-exercise"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-bonus-exercise-results.html"><a href="module-1-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 1: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-1.-preparing-our-data"><i class="fa fa-check"></i>Step 1. Preparing Our Data</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-2.-apply-pca"><i class="fa fa-check"></i>Step 2. Apply PCA</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-3.-visualisation-of-pca-results"><i class="fa fa-check"></i>Step 3. Visualisation of PCA results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#bonus-exercise-1"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-bonus-exercise-results.html"><a href="module-2-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 2: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#bonus-exercise-2"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-bonus-exercise-results.html"><a href="module-3-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 3: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes from RNAseq data</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#bonus-exercise-3"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-bonus-exercise-results.html"><a href="module-4-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 4: Bonus Exercise Results</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Analysis Using R</h1>
<h2 class="subtitle"><em>Canadian Bioinformatics Workshop</em></h2>
<p class="author"><em>Instructors: Shraddha Pai, Chaitra Sarathy</em></p>
<p class="date"><em>last modified 2024-06-10</em></p>
</div>
<div id="welcome" class="section level1 unnumbered hasAnchor">
<h1>Welcome<a href="index.html#welcome" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Welcome to <a href="https://bioinformatics.ca/workshops-all/2024-analysis-using-r/">Analysis Using R</a> 2024.</p>
<div id="meet-your-faculty" class="section level2 unnumbered hasAnchor">
<h2>Meet your Faculty<a href="index.html#meet-your-faculty" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="shraddha-pai" class="section level3 unlisted unnumbered hasAnchor">
<h3>Shraddha Pai<a href="index.html#shraddha-pai" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Investigator I, OICR<br>
Assistant Professor, University of Toronto</p>
<p><img src="images/ShraddhaPai.jpeg" width=200 height=150 ALIGN="right"  hspace=25>
Dr. Pai integrates genomics and computational methods to advance precision medicine. Her previous work involves DNA methylome-based biomarker discovery in psychosis, and building machine learning algorithms for patient classification from multi-modal data. The <a href="https://www.shraddhapailab.com/">Pai Lab at the Ontario Institute for Cancer Research</a> focuses on biomarker discovery for detection, diagnosis and prognosis in brain cancers and other brain-related disorders.</p>
</div>
<div id="chaitra-sarathy" class="section level3 unlisted unnumbered hasAnchor">
<h3>Chaitra Sarathy<a href="index.html#chaitra-sarathy" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Bioinformatics Specialist <br>
Krembil Research Institute<br></p>
<p><img src="images/Chaitra_Sarathy.jpeg" ALIGN="right" hspace=25>
Dr. Sarathy is a computational biologist with industry experience in software development.
Her previous research centered around developing multi-scale mathematical models of human systems to characterise biochemical changes in obesity. In addition, she has developed methods based on machine learning and multi-omics integration to identify drug targets in cancer and stratify patients for clinical trials. She currently focusses on characterising genetic malfunctions in neurological diseases.</p>
</div>
<div id="ian-cheong" class="section level3 unlisted unnumbered hasAnchor">
<h3>Ian Cheong<a href="index.html#ian-cheong" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/IanCheong.jpg" ALIGN="right" hspace=25 width=200 height=200 style="object-fit: cover;">
MSc. Candidate<br>
University of Toronto</p>
<p>Ian is a Master’s level candidate in the Department of Medical Biophysics at the University of Toronto. His thesis work in the Pai lab involves analysis of single-cell transcriptomes to find the link between brain development and the development of childhood brain cancer.
  <br>
  <br></p>
</div>
<div id="zoe-klein" class="section level3 unlisted unnumbered hasAnchor">
<h3>Zoe Klein<a href="index.html#zoe-klein" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/ZoeKlein.png" ALIGN="right" hspace=25 width=200 height=200 style="object-fit: cover;">
PhD. Candidate<br>
University of Toronto</p>
<p>Zoe is a PhD level candidate in the Department of Molecular Genetics at the University of Toronto. Her thesis work in the Reimand lab involves using computational tools to investigate the role of non-coding RNA in cancer.
  <br>
  <br></p>
</div>
<div id="amin-nooranikhojasteh" class="section level3 unlisted unnumbered hasAnchor">
<h3>Amin Nooranikhojasteh<a href="index.html#amin-nooranikhojasteh" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="nia-hughes" class="section level3 unlisted unnumbered hasAnchor">
<h3>Nia Hughes<a href="index.html#nia-hughes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/nia_hughes.jpeg" ALIGN="right" hspace=25>
Program Manager<br>
Bioinformatics.ca<br>
<a href="mailto:nia.hughes@oicr.on.ca">nia.hughes@oicr.on.ca</a></p>
<p>Nia is the Program Manager for Bioinformatics.ca, where she coordinates the Canadian Bioinformatics Workshop Series. Prior to starting at the OICR, she completed her M.Sc in Bioinformatics from the University of Guelph in 2020 before working there as a bioinformatician studying epigenetic and transcriptomic patterns across maize varieties.</p>
<p>
</p>
</div>
</div>
<div id="pre-workshop-materials-and-laptop-setup-instructions" class="section level2 unnumbered hasAnchor">
<h2>Pre-workshop Materials and Laptop Setup Instructions<a href="index.html#pre-workshop-materials-and-laptop-setup-instructions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="laptop-setup-instructions" class="section level3 unnumbered hasAnchor">
<h3>Laptop Setup Instructions<a href="index.html#laptop-setup-instructions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A checklist to setup your laptop can be found here.</p>
<p>Install these tools on your laptop before coming to the workshop:</p>
<ol style="list-style-type: decimal">
<li>R (4.0+)
<i>Note:</i> MacBook users with an Apple silicon chip (e.g., M1 or M2) should install the “arm64” version of R, while MacBook users with an Intel chip should install the regular (64-bit) version of R. You can check your laptop’s hardware specifications by clicking the Apple icon (top left corner) &gt; About This Mac and verifying whether the chip is Apple or Intel.</li>
<li>Rstudio</li>
<li>Make sure you have a robust internet browser such as Firefox, Safari or Chrome (not Internet Explorer).</li>
<li>Make sure you have a PDF viewer (e.g. Adobe Acrobat, Preview or similar) or that you can read PDF files in your Web browser.</li>
</ol>
</div>
<div id="r-packages" class="section level3 unnumbered hasAnchor">
<h3>R packages<a href="index.html#r-packages" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" tabindex="-1"></a>pkgList <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;tidyverse&quot;</span>, <span class="st">&quot;clValid&quot;</span>,<span class="st">&quot;rgl&quot;</span>,<span class="st">&quot;RColorBrewer&quot;</span>,<span class="st">&quot;corrplot&quot;</span>,<span class="st">&quot;ClusterR&quot;</span>,</span>
<span id="cb1-2"><a href="index.html#cb1-2" tabindex="-1"></a>  <span class="st">&quot;Rtsne&quot;</span>,<span class="st">&quot;umap&quot;</span>,<span class="st">&quot;BiocManager&quot;</span>,<span class="st">&quot;mlbench&quot;</span>,<span class="st">&quot;plotrix&quot;</span>, <span class="st">&quot;factoextra&quot;</span>)</span>
<span id="cb1-3"><a href="index.html#cb1-3" tabindex="-1"></a><span class="cf">for</span> (cur <span class="cf">in</span> pkgList){</span>
<span id="cb1-4"><a href="index.html#cb1-4" tabindex="-1"></a>  <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">&quot;</span><span class="sc">\t</span><span class="st">Checking for %s ...&quot;</span>, cur))</span>
<span id="cb1-5"><a href="index.html#cb1-5" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(cur, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(cur)</span>
<span id="cb1-6"><a href="index.html#cb1-6" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>##  Checking for tidyverse ...</code></pre>
<pre><code>##  Checking for clValid ...</code></pre>
<pre><code>##  Checking for rgl ...</code></pre>
<pre><code>##  Checking for RColorBrewer ...</code></pre>
<pre><code>##  Checking for corrplot ...</code></pre>
<pre><code>##  Checking for ClusterR ...</code></pre>
<pre><code>##  Checking for Rtsne ...</code></pre>
<pre><code>##  Checking for umap ...</code></pre>
<pre><code>##  Checking for BiocManager ...</code></pre>
<pre><code>##  Checking for mlbench ...</code></pre>
<pre><code>##  Checking for plotrix ...</code></pre>
<pre><code>##  Checking for factoextra ...</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="index.html#cb14-1" tabindex="-1"></a>biocPkg <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;edgeR&quot;</span>,<span class="st">&quot;bladderbatch&quot;</span>,<span class="st">&quot;airway&quot;</span>)</span>
<span id="cb14-2"><a href="index.html#cb14-2" tabindex="-1"></a><span class="cf">for</span> (cur <span class="cf">in</span> biocPkg){</span>
<span id="cb14-3"><a href="index.html#cb14-3" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(cur, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) BiocManager<span class="sc">::</span><span class="fu">install</span>(cur)</span>
<span id="cb14-4"><a href="index.html#cb14-4" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="download-example-data" class="section level3 unnumbered hasAnchor">
<h3>Download example data<a href="index.html#download-example-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ol style="list-style-type: decimal">
<li>Download data for the workshop: <a href="https://drive.google.com/file/d/1DUSqDM1DGQaqPOk1iL8JxfU0TgCr-4MR/view?usp=sharing">AUR2024_data.zip</a>.</li>
<li>Move the zip file to your working directory (usually under “/Users/yourname”).</li>
<li>Unzip the file. You should see a folder with 4 files in it.</li>
</ol>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="lecture-slides.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
