<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 4: Finding differentially expressed genes from RNAseq data | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 4: Finding differentially expressed genes from RNAseq data | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 4: Finding differentially expressed genes from RNAseq data | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-3-generalized-linear-models.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2024 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-1.-preparing-our-data"><i class="fa fa-check"></i>Step 1. Preparing Our Data</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-2.-apply-pca"><i class="fa fa-check"></i>Step 2. Apply PCA</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-3.-visualisation-of-pca-results"><i class="fa fa-check"></i>Step 3. Visualisation of PCA results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes from RNAseq data</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-4-finding-differentially-expressed-genes-from-rnaseq-data" class="section level1 unnumbered hasAnchor">
<h1>Module 4: Finding differentially expressed genes from RNAseq data<a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#module-4-finding-differentially-expressed-genes-from-rnaseq-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this exercise, we will use <code>edgeR</code> to call differentially-expressed genes. For the example we will use an RNAseq dataset from a treatment-vehicle design.</p>
<p>If you haven’t already done so, install the <code>airway</code> dataset:</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb204-1" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;airway&quot;) </span></span></code></pre></div>
<p>These data are from the paper:</p>
<p>Himes BE, Jiang X, Wagner P, Hu R, Wang Q, Klanderman B, Whitaker RM, Duan Q, Lasky-Su J, Nikolos C, Jester W, Johnson M, Panettieri R Jr, Tantisira KG, Weiss ST, Lu Q. “RNA-Seq Transcriptome Profiling Identifies CRISPLD2 as a Glucocorticoid Responsive Gene that Modulates Cytokine Function in Airway Smooth Muscle Cells.” PLoS One. 2014 Jun 13;9(6):e99625. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/24926665/">24926665</a>.</p>
<p>From the abstract of the original paper:
“Using RNA-Seq, a high-throughput sequencing method, we characterized transcriptomic changes in four primary human ASM cell lines that were treated with dexamethasone - a potent synthetic glucocorticoid (1 micromolar for 18 hours).”</p>
<p>Let’s load the data</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb205-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(airway))</span>
<span id="cb205-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb205-2" tabindex="-1"></a><span class="fu">data</span>(airway)</span></code></pre></div>
<p>These data happen to be in a Bioconductor-specific format, so we use the special functions, <code>assay()</code> and <code>colData()</code> to get the expression data and smaple information.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb206-1" tabindex="-1"></a>xpr <span class="ot">&lt;-</span> <span class="fu">assay</span>(airway)</span>
<span id="cb206-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb206-2" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">colData</span>(airway)</span></code></pre></div>
<p><strong>Questions:</strong><br />
* How many samples are in this experiment?
* How many genes were measured?
* How many treatment groups are there (<code>dex</code> column)?</p>
<p>Let’s created a <code>DGEList</code> object (DGE stands for “Differential Gene Expression”). This object is what we will use for our differential expression analysis.</p>
<p>Note: Make phenotype of interest categorical. In R that means converting to a <code>factor</code> type with categorical <code>levels</code>. You can think of levels as ordinal representations (e.g., first level = 1, second = 2, etc., )</p>
<p>If <code>levels=</code> are not set, the default uses alphabetical order. We recommend explicitly setting levels so that there are no assumptions.</p>
<p>Load the <code>edgeR</code> package:</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb207-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb207-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(edgeR))</span></code></pre></div>
<p>Let’s create a <code>DGEList</code> object for the differential expression analysis. Note that <code>group</code> must be a categorical variable (use <code>factor()</code> to convert it to one):</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb208-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb208-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(samples<span class="sc">$</span>dex)</span>
<span id="cb208-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb208-2" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb208-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb208-3" tabindex="-1"></a>    <span class="at">counts =</span> xpr,</span>
<span id="cb208-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb208-4" tabindex="-1"></a>    <span class="at">group =</span> group</span>
<span id="cb208-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb208-5" tabindex="-1"></a>    )</span></code></pre></div>
<p>Remove low-count genes: To filter low count genes, we’re going to use a normalized count measure called <code>cpm</code> (counts per million). We are going to keep genes with 100 or greater counts per million for at least two samples:</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb209-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb209-1" tabindex="-1"></a><span class="fu">head</span>(dge<span class="sc">$</span>counts)</span></code></pre></div>
<pre><code>##                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516 SRR1039517 SRR1039520
## ENSG00000000003        679        448        873        408       1138       1047        770
## ENSG00000000005          0          0          0          0          0          0          0
## ENSG00000000419        467        515        621        365        587        799        417
## ENSG00000000457        260        211        263        164        245        331        233
## ENSG00000000460         60         55         40         35         78         63         76
## ENSG00000000938          0          0          2          0          1          0          0
##                 SRR1039521
## ENSG00000000003        572
## ENSG00000000005          0
## ENSG00000000419        508
## ENSG00000000457        229
## ENSG00000000460         60
## ENSG00000000938          0</code></pre>
<p>Look at counts per million using <code>cpm</code>:</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb211-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb211-1" tabindex="-1"></a><span class="fu">cpm</span>(dge)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516
## ENSG00000000003  32.900521  23.817776  34.439705  26.906868  46.546998
## ENSG00000000005   0.000000   0.000000   0.000000   0.000000   0.000000
## ENSG00000000419  22.628193  27.379809  24.498347  24.071095  24.009743
## ENSG00000000457  12.598138  11.217747  10.375306  10.815506  10.021102
## ENSG00000000460   2.907263   2.924057   1.577993   2.308187   3.190392</code></pre>
<p>This next line is a bit complex so let’s unpack it:</p>
<ul>
<li>We are using <code>cpm(dge)&gt;100</code> as a logical test (“which genes have cpm &gt; 100?”).</li>
<li>For each gene, we want that test to be true for at least two samples. For this we use <code>rowSums()</code> to add up how many samples meet that criteria.</li>
</ul>
<div class="sourceCode" id="cb213"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb213-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb213-1" tabindex="-1"></a><span class="fu">dim</span>(dge) <span class="co">#before </span></span></code></pre></div>
<pre><code>## [1] 63677     8</code></pre>
<div class="sourceCode" id="cb215"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb215-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-1" tabindex="-1"></a><span class="co"># keep genes which have cpm&gt;100 in 2 or more samples</span></span>
<span id="cb215-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-2" tabindex="-1"></a>tokeep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">cpm</span>(dge)<span class="sc">&gt;</span><span class="dv">100</span>) <span class="sc">&gt;=</span> <span class="dv">2</span> </span>
<span id="cb215-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-3" tabindex="-1"></a></span>
<span id="cb215-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-4" tabindex="-1"></a><span class="co"># now filter for these</span></span>
<span id="cb215-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-5" tabindex="-1"></a>dge <span class="ot">&lt;-</span> dge[tokeep,keep.lib.sizes <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb215-6"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-6" tabindex="-1"></a></span>
<span id="cb215-7"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-7" tabindex="-1"></a><span class="co"># how many genes do we have left?</span></span>
<span id="cb215-8"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb215-8" tabindex="-1"></a><span class="fu">dim</span>(dge) <span class="co">#after</span></span></code></pre></div>
<pre><code>## [1] 2086    8</code></pre>
<p>Normalize the data:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb217-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb217-1" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge)</span></code></pre></div>
<p>Visualize the data:</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb218-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-1" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb218-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-2" tabindex="-1"></a>    dge, </span>
<span id="cb218-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-3" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">as.numeric</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>group), </span>
<span id="cb218-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-4" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb218-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-5" tabindex="-1"></a>)</span>
<span id="cb218-6"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-6" tabindex="-1"></a><span class="fu">legend</span>(</span>
<span id="cb218-7"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-7" tabindex="-1"></a>    <span class="st">&quot;bottomleft&quot;</span>, </span>
<span id="cb218-8"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-8" tabindex="-1"></a>    <span class="fu">as.character</span>(<span class="fu">unique</span>(dge<span class="sc">$</span>samples<span class="sc">$</span>group)),</span>
<span id="cb218-9"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-9" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb218-10"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb218-10" tabindex="-1"></a>    )</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-64-1.png" width="672" /></p>
<p>Let’s create a model design to identify genes with a <code>group</code> effect:</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb219-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb219-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples<span class="sc">$</span>group</span>
<span id="cb219-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb219-2" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>group)</span></code></pre></div>
<p>Estimate variation (“dispersion”) for each gene:</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb220-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb220-1" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">estimateDisp</span>(dge, mod)</span></code></pre></div>
<p>Call differentially expressed genes.</p>
<p>Here we:</p>
<ul>
<li>fit a model for each gene, using <code>glmFit</code></li>
<li>we have built in an estimate of gene-wise dispersion to better identify treatment effect (or “contrast”)</li>
<li>for each gene, we run a likelihood ratio test which compares which model fits the data better: a null model (treatment effect = 0)
or a full model (treatment effect is non-zero)</li>
</ul>
<p>Note that <code>coef=2</code> fetches the effects for the treatment effect; <code>coef=1</code> would fetch effects of the intercept term.<br>
  <br>
<img src="images/model_coefs.png" width="75%"></p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb221-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb221-1" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">glmFit</span>(dge,mod)</span>
<span id="cb221-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb221-2" tabindex="-1"></a>diffEx <span class="ot">&lt;-</span> <span class="fu">glmLRT</span>(fit, <span class="at">coef =</span> <span class="dv">2</span>) <span class="co"># get coefficients for group term</span></span></code></pre></div>
<p>Look at the top 10 differentially expressed genes:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb222-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb222-1" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">topTags</span>(diffEx, <span class="at">n=</span><span class="dv">10</span>)</span>
<span id="cb222-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb222-2" tabindex="-1"></a>tt</span></code></pre></div>
<pre><code>## Coefficient:  groupuntrt 
##                     logFC   logCPM       LR       PValue          FDR
## ENSG00000152583 -4.512108 5.950547 190.5712 2.385925e-43 4.977039e-40
## ENSG00000178695  2.592010 7.433587 171.7187 3.117449e-39 3.251499e-36
## ENSG00000120129 -2.857535 7.727638 167.2022 3.021823e-38 2.101174e-35
## ENSG00000189221 -3.213455 7.183776 162.3746 3.426477e-37 1.786908e-34
## ENSG00000125148 -2.110664 7.835949 156.1280 7.937896e-36 3.311690e-33
## ENSG00000162614 -1.938280 8.397283 135.2052 2.977497e-31 1.035176e-28
## ENSG00000101347 -3.764745 9.620058 128.9686 6.889897e-30 2.053189e-27
## ENSG00000096060 -3.849662 7.313033 123.8841 8.930961e-29 2.328748e-26
## ENSG00000134686 -1.294763 7.426574 121.0663 3.695702e-28 8.565816e-26
## ENSG00000166741 -2.072008 8.859215 114.0478 1.272045e-26 2.653486e-24</code></pre>
<p>For the next steps we’re going to need stats on all the genes we’ve tested. So let’s get those:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb224-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb224-1" tabindex="-1"></a>tt <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(</span>
<span id="cb224-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb224-2" tabindex="-1"></a>    <span class="fu">topTags</span>(diffEx, <span class="at">n=</span><span class="fu">nrow</span>(dge)</span>
<span id="cb224-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb224-3" tabindex="-1"></a>    )</span>
<span id="cb224-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb224-4" tabindex="-1"></a>)</span></code></pre></div>
<p>A QQplot directly compares the pvalues from our statistical tests to the expected values from a random uniform distribution (p-value selected at random).</p>
<p>A deviation from the x=y line (diagonal) towards the top indicates an enrichment of signal.</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb225-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-1" tabindex="-1"></a><span class="fu">qqplot</span>(</span>
<span id="cb225-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-2" tabindex="-1"></a>    tt<span class="sc">$</span>PValue, </span>
<span id="cb225-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-3" tabindex="-1"></a>    <span class="fu">runif</span>(<span class="fu">nrow</span>(tt)), <span class="co"># randomly sample from uniform distribution</span></span>
<span id="cb225-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-4" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;p-values from real data&quot;</span>,</span>
<span id="cb225-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-5" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Randomly-sampled values from Uniform distribution&quot;</span>,</span>
<span id="cb225-6"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-6" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>,<span class="at">cex=</span><span class="fl">0.5</span></span>
<span id="cb225-7"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-7" tabindex="-1"></a>)</span>
<span id="cb225-8"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-8" tabindex="-1"></a></span>
<span id="cb225-9"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-9" tabindex="-1"></a><span class="co"># x=y line as reference</span></span>
<span id="cb225-10"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb225-10" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/qqplot-1.png" width="672" /></p>
<p>Now let’s call differentially expressed genes using the <code>decideTestDGE()</code> function and use <code>summary()</code> to see how many genes are upregulated (value <code>+1</code>), downregulated (value <code>-1</code>) and not called as changed (value <code>0</code>)</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb226-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb226-1" tabindex="-1"></a>diffEx2 <span class="ot">&lt;-</span> <span class="fu">decideTestsDGE</span>(diffEx, </span>
<span id="cb226-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb226-2" tabindex="-1"></a>    <span class="at">adjust.method=</span><span class="st">&quot;BH&quot;</span>, </span>
<span id="cb226-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb226-3" tabindex="-1"></a>    <span class="at">p.value=</span><span class="fl">0.05</span></span>
<span id="cb226-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb226-4" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Warning in decideTestsDGE(diffEx, adjust.method = &quot;BH&quot;, p.value = 0.05): &#39;decideTestsDGE&#39; is deprecated.
## Use &#39;decideTests&#39; instead.
## See help(&quot;Deprecated&quot;)</code></pre>
<div class="sourceCode" id="cb228"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb228-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb228-1" tabindex="-1"></a><span class="fu">summary</span>(diffEx2)</span></code></pre></div>
<pre><code>##        groupuntrt
## Down          317
## NotSig       1553
## Up            216</code></pre>
<p>A <strong>volcano plot</strong> can help visualize effect magnitude - log2 fold-change or <code>log2FC</code> in the table ` against the corresponding p-value. Here we create a volcano plot, and colour-code upregulated genes in red, and downregulated genes in blue.</p>
<p>Note that we are combining two different tables, <code>tt</code> and <code>diffEx2</code> so we need to ensure the order is the same. Otherwise the colours will be in the wrong order <em>(try it for yourself!)</em>.</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb230-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-1" tabindex="-1"></a>midx <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">rownames</span>(tt), <span class="fu">rownames</span>(diffEx2))</span>
<span id="cb230-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-2" tabindex="-1"></a>diffEx2 <span class="ot">&lt;-</span> diffEx2[midx,]</span>
<span id="cb230-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-3" tabindex="-1"></a></span>
<span id="cb230-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-4" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">&quot;black&quot;</span>,<span class="fu">nrow</span>(diffEx2))</span>
<span id="cb230-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-5" tabindex="-1"></a>cols[<span class="fu">which</span>(diffEx2<span class="sc">&gt;</span><span class="dv">0</span> )]  <span class="ot">&lt;-</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb230-6"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-6" tabindex="-1"></a>cols[<span class="fu">which</span>(diffEx2<span class="sc">&lt;</span><span class="dv">0</span>)]  <span class="ot">&lt;-</span> <span class="st">&quot;blue&quot;</span></span>
<span id="cb230-7"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-7" tabindex="-1"></a><span class="co"># volcano plot</span></span>
<span id="cb230-8"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-8" tabindex="-1"></a><span class="fu">plot</span>(tt<span class="sc">$</span>logFC,<span class="sc">-</span><span class="fu">log10</span>(tt<span class="sc">$</span>PValue),<span class="at">pch=</span><span class="dv">16</span>,</span>
<span id="cb230-9"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-9" tabindex="-1"></a>    <span class="at">col=</span>cols)</span>
<span id="cb230-10"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb230-10" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="dv">0</span>,<span class="at">lty=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-65-1.png" width="672" /></p>
<p>Finally we can write our differential expression results out to file:</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb231-1"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-1" tabindex="-1"></a><span class="fu">write.table</span>(tt,<span class="at">file=</span><span class="st">&quot;diffEx.results.txt&quot;</span>,</span>
<span id="cb231-2"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-2" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,</span>
<span id="cb231-3"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-3" tabindex="-1"></a>    <span class="at">col=</span><span class="cn">TRUE</span>,</span>
<span id="cb231-4"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-4" tabindex="-1"></a>    <span class="at">row=</span><span class="cn">TRUE</span>,</span>
<span id="cb231-5"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-5" tabindex="-1"></a>    <span class="at">quote=</span><span class="cn">FALSE</span></span>
<span id="cb231-6"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#cb231-6" tabindex="-1"></a>)</span></code></pre></div>


</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-3-generalized-linear-models.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
