<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3: Bonus Exercise Results | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3: Bonus Exercise Results | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3: Bonus Exercise Results | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-3-generalized-linear-models.html"/>
<link rel="next" href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2024 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#download-example-data"><i class="fa fa-check"></i>Download example data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#bonus-exercise"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-bonus-exercise-results.html"><a href="module-1-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 1: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-1.-preparing-our-data"><i class="fa fa-check"></i>Step 1. Preparing Our Data</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-2.-apply-pca"><i class="fa fa-check"></i>Step 2. Apply PCA</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-3.-visualisation-of-pca-results"><i class="fa fa-check"></i>Step 3. Visualisation of PCA results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#bonus-exercise-1"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-bonus-exercise-results.html"><a href="module-2-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 2: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#bonus-exercise-2"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-bonus-exercise-results.html"><a href="module-3-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 3: Bonus Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes from RNAseq data</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#volcano-plot-r-base-graphics"><i class="fa fa-check"></i>Volcano plot (R base graphics)</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#volcano-plot-ggplot2"><i class="fa fa-check"></i>Volcano plot (ggplot2)</a></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html#bonus-exercise-3"><i class="fa fa-check"></i>Bonus Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-bonus-exercise-results.html"><a href="module-4-bonus-exercise-results.html"><i class="fa fa-check"></i>Module 4: Bonus Exercise Results</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3-bonus-exercise-results" class="section level1 unnumbered hasAnchor">
<h1>Module 3: Bonus Exercise Results<a href="module-3-bonus-exercise-results.html#module-3-bonus-exercise-results" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div class="sourceCode" id="cb272"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb272-1"><a href="module-3-bonus-exercise-results.html#cb272-1" tabindex="-1"></a><span class="fu">data</span>(BreastCancer)</span>
<span id="cb272-2"><a href="module-3-bonus-exercise-results.html#cb272-2" tabindex="-1"></a>bc <span class="ot">&lt;-</span> BreastCancer</span>
<span id="cb272-3"><a href="module-3-bonus-exercise-results.html#cb272-3" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>) <span class="co"># altered for current lab</span></span>
<span id="cb272-4"><a href="module-3-bonus-exercise-results.html#cb272-4" tabindex="-1"></a>    bc[,k] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(bc[,k]) </span>
<span id="cb272-5"><a href="module-3-bonus-exercise-results.html#cb272-5" tabindex="-1"></a><span class="fu">head</span>(bc)</span></code></pre></div>
<pre><code>##        Id Cl.thickness
## 1 1000025            5
## 2 1002945            5
## 3 1015425            3
## 4 1016277            6
## 5 1017023            4
## 6 1017122            8
##   Cell.size Cell.shape
## 1         1          1
## 2         4          4
## 3         1          1
## 4         8          8
## 5         1          1
## 6        10         10
##   Marg.adhesion Epith.c.size
## 1             1            2
## 2             5            7
## 3             1            2
## 4             1            3
## 5             3            2
## 6             8            7
##   Bare.nuclei Bl.cromatin
## 1           1           3
## 2          10           3
## 3           2           3
## 4           4           3
## 5           1           3
## 6          10           9
##   Normal.nucleoli Mitoses
## 1               1       1
## 2               2       1
## 3               1       1
## 4               7       1
## 5               1       1
## 6               7       1
##       Class
## 1    benign
## 2    benign
## 3    benign
## 4    benign
## 5    benign
## 6 malignant</code></pre>
<p>Explore missingness:</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb274-1"><a href="module-3-bonus-exercise-results.html#cb274-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(plotrix))</span>
<span id="cb274-2"><a href="module-3-bonus-exercise-results.html#cb274-2" tabindex="-1"></a></span>
<span id="cb274-3"><a href="module-3-bonus-exercise-results.html#cb274-3" tabindex="-1"></a><span class="co">#&#39; show data missingness as a chequered matrix</span></span>
<span id="cb274-4"><a href="module-3-bonus-exercise-results.html#cb274-4" tabindex="-1"></a><span class="co">#&#39; </span></span>
<span id="cb274-5"><a href="module-3-bonus-exercise-results.html#cb274-5" tabindex="-1"></a><span class="co">#&#39; @param x (matrix) data matrix.</span></span>
<span id="cb274-6"><a href="module-3-bonus-exercise-results.html#cb274-6" tabindex="-1"></a><span class="co">#&#39; @param outFile (char) path to file for printing graph</span></span>
<span id="cb274-7"><a href="module-3-bonus-exercise-results.html#cb274-7" tabindex="-1"></a><span class="co">#&#39; @param wd (numeric) width in inches</span></span>
<span id="cb274-8"><a href="module-3-bonus-exercise-results.html#cb274-8" tabindex="-1"></a><span class="co">#&#39; @param ht (numeric) height in inches</span></span>
<span id="cb274-9"><a href="module-3-bonus-exercise-results.html#cb274-9" tabindex="-1"></a><span class="co">#&#39; @return plots missingness matrix to file</span></span>
<span id="cb274-10"><a href="module-3-bonus-exercise-results.html#cb274-10" tabindex="-1"></a><span class="co">#&#39; @import plotrix</span></span>
<span id="cb274-11"><a href="module-3-bonus-exercise-results.html#cb274-11" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb274-12"><a href="module-3-bonus-exercise-results.html#cb274-12" tabindex="-1"></a>plotMissMat <span class="ot">&lt;-</span> <span class="cf">function</span>(x,<span class="at">xlab=</span><span class="st">&quot;columns&quot;</span>,</span>
<span id="cb274-13"><a href="module-3-bonus-exercise-results.html#cb274-13" tabindex="-1"></a>        <span class="at">ylab=</span><span class="st">&quot;rows&quot;</span>,<span class="at">border=</span><span class="cn">NA</span>) {</span>
<span id="cb274-14"><a href="module-3-bonus-exercise-results.html#cb274-14" tabindex="-1"></a>    </span>
<span id="cb274-15"><a href="module-3-bonus-exercise-results.html#cb274-15" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(x)</span>
<span id="cb274-16"><a href="module-3-bonus-exercise-results.html#cb274-16" tabindex="-1"></a>    <span class="fu">class</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;numeric&quot;</span></span>
<span id="cb274-17"><a href="module-3-bonus-exercise-results.html#cb274-17" tabindex="-1"></a>    <span class="fu">color2D.matplot</span>(x,<span class="at">show.values=</span><span class="cn">FALSE</span>,<span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb274-18"><a href="module-3-bonus-exercise-results.html#cb274-18" tabindex="-1"></a>        <span class="at">cs1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs2=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs3=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">border=</span>border,</span>
<span id="cb274-19"><a href="module-3-bonus-exercise-results.html#cb274-19" tabindex="-1"></a>        <span class="at">cex=</span><span class="fl">0.8</span>,</span>
<span id="cb274-20"><a href="module-3-bonus-exercise-results.html#cb274-20" tabindex="-1"></a>        <span class="at">xlab=</span>xlab,<span class="at">ylab=</span>ylab)</span>
<span id="cb274-21"><a href="module-3-bonus-exercise-results.html#cb274-21" tabindex="-1"></a>}</span></code></pre></div>
<p>Explore missingness:</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb275-1"><a href="module-3-bonus-exercise-results.html#cb275-1" tabindex="-1"></a><span class="fu">plotMissMat</span>(bc)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/ex3-miss2-1.png" width="672" /></p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb276-1"><a href="module-3-bonus-exercise-results.html#cb276-1" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(bc))</span></code></pre></div>
<pre><code>##              Id 
##               0 
##    Cl.thickness 
##               0 
##       Cell.size 
##               0 
##      Cell.shape 
##               0 
##   Marg.adhesion 
##               0 
##    Epith.c.size 
##               0 
##     Bare.nuclei 
##              16 
##     Bl.cromatin 
##               0 
## Normal.nucleoli 
##               0 
##         Mitoses 
##               0 
##           Class 
##               0</code></pre>
<p>Plot relationship between variables</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb278-1"><a href="module-3-bonus-exercise-results.html#cb278-1" tabindex="-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb278-2"><a href="module-3-bonus-exercise-results.html#cb278-2" tabindex="-1"></a></span>
<span id="cb278-3"><a href="module-3-bonus-exercise-results.html#cb278-3" tabindex="-1"></a><span class="fu">ggplot</span>(bc,<span class="fu">aes</span>(Cell.size,Normal.nucleoli)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/ex3-plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb279-1"><a href="module-3-bonus-exercise-results.html#cb279-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Normal.nucleoli))</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/ex3-plot-2.png" width="672" /></p>
<div class="sourceCode" id="cb280"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb280-1"><a href="module-3-bonus-exercise-results.html#cb280-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Cl.thickness))</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/ex3-plot-3.png" width="672" /></p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb281-1"><a href="module-3-bonus-exercise-results.html#cb281-1" tabindex="-1"></a><span class="fu">ggplot</span>(bc) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="fu">factor</span>(Class), Bare.nuclei))</span></code></pre></div>
<pre><code>## Warning: Removed 16 rows containing
## non-finite outside the scale
## range (`stat_boxplot()`).</code></pre>
<p><img src="bookdown-demo_files/figure-html/ex3-plot-4.png" width="672" /></p>
<p>Fit a binary outcome model:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb283-1"><a href="module-3-bonus-exercise-results.html#cb283-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(</span>
<span id="cb283-2"><a href="module-3-bonus-exercise-results.html#cb283-2" tabindex="-1"></a>    Class <span class="sc">~</span> Cl.thickness <span class="sc">+</span> Bare.nuclei <span class="sc">+</span> Normal.nucleoli <span class="sc">+</span> Mitoses <span class="sc">+</span> Bl.cromatin, </span>
<span id="cb283-3"><a href="module-3-bonus-exercise-results.html#cb283-3" tabindex="-1"></a>    bc,</span>
<span id="cb283-4"><a href="module-3-bonus-exercise-results.html#cb283-4" tabindex="-1"></a>    <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb283-5"><a href="module-3-bonus-exercise-results.html#cb283-5" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = Class ~ Cl.thickness + Bare.nuclei + Normal.nucleoli + 
##     Mitoses + Bl.cromatin, family = &quot;binomial&quot;, data = bc)
## 
## Coefficients:
##                 Estimate
## (Intercept)     -9.69635
## Cl.thickness     0.63413
## Bare.nuclei      0.50988
## Normal.nucleoli  0.35765
## Mitoses          0.52677
## Bl.cromatin      0.61398
##                 Std. Error
## (Intercept)        0.98763
## Cl.thickness       0.12267
## Bare.nuclei        0.08228
## Normal.nucleoli    0.10040
## Mitoses            0.27923
## Bl.cromatin        0.14225
##                 z value
## (Intercept)      -9.818
## Cl.thickness      5.169
## Bare.nuclei       6.197
## Normal.nucleoli   3.562
## Mitoses           1.887
## Bl.cromatin       4.316
##                 Pr(&gt;|z|)    
## (Intercept)      &lt; 2e-16 ***
## Cl.thickness    2.35e-07 ***
## Bare.nuclei     5.76e-10 ***
## Normal.nucleoli 0.000368 ***
## Mitoses         0.059223 .  
## Bl.cromatin     1.59e-05 ***
## ---
## Signif. codes:  
##   0 &#39;***&#39; 0.001 &#39;**&#39; 0.01
##   &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 884.35  on 682  degrees of freedom
## Residual deviance: 120.81  on 677  degrees of freedom
##   (16 observations deleted due to missingness)
## AIC: 132.81
## 
## Number of Fisher Scoring iterations: 8</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-3-generalized-linear-models.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
