--- 
title: "Analysis Using R"
subtitle: "Canadian Bioinformatics Workshop"
author: "Instructors: Shraddha Pai, Chaitra Sarathy"
date: "last modified `r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib]
biblio-style: nature
csl: nature.csl
link-citations: yes
github-repo: rstudio/bookdown-demo
favicon: images/favicon.ico
description: "Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models"
---

# Welcome {-}
Welcome to [Analysis Using R](https://bioinformatics.ca/workshops-all/2024-analysis-using-r/) 2024.

## Meet your Faculty {-}

### Shraddha Pai {.unlisted .unnumbered}
Investigator I, OICR<br>
Assistant Professor, University of Toronto

<img src="images/ShraddhaPai.jpeg" width=200 height=150 ALIGN="right"  hspace=25>
Dr. Pai integrates genomics and computational methods to advance precision medicine. Her previous work involves DNA methylome-based biomarker discovery in psychosis, and building machine learning algorithms for patient classification from multi-modal data. The <a href="https://www.shraddhapailab.com/">Pai Lab at the Ontario Institute for Cancer Research</a> focuses on biomarker discovery for detection, diagnosis and prognosis in brain cancers and other brain-related disorders. 

### Chaitra Sarathy {.unlisted .unnumbered}
Scientific Associate <br>
TBA<br>
TBA

<img src="images/Chaitra_Sarathy.jpeg" ALIGN="right" hspace=25>
Dr. Sarathy is a computational biologist with industry experience in software development. 
Her previous research centered around developing multi-scale mathematical models of human systems to characterise biochemical changes in obesity. In addition, she has developed methods based on machine learning and multi-omics integration to identify drug targets in cancer and stratify patients for clinical trials. She is currently a Bioinformatics Specialist at Krembil Research Institute where she focusses on characterising genetic malfunctions in neurological diseases.

### Ian Cheong {.unlisted .unnumbered}
<img src="images/IanCheong.jpg" ALIGN="right" hspace=25 width=200 height=200 style="object-fit: cover;">
MSc. Candidate<br>
University of Toronto

Ian is a Master's level candidate in the Department of Medical Biophysics at the University of Toronto. His thesis work in the Pai lab involves analysis of single-cell transcriptomes to find the link between brain development and the development of childhood brain cancer. 
&nbsp;&nbsp;<br>
&nbsp;&nbsp;<br>

### Zoe Klein {.unlisted .unnumbered}
PhD. Candidate<br>
University of Toronto

Zoe is a PhD level candidate in the Department of Molecular Genetics at the University of Toronto. Her thesis work in the Reimand lab involves using computational tools to investigate the role of non-coding RNA in cancer. 
&nbsp;&nbsp;<br>
&nbsp;&nbsp;<br>

### Nia Hughes {.unlisted .unnumbered}

<img src="images/nia_hughes.jpeg" ALIGN="right" hspace=25>
Program Manager<br>
Bioinformatics.ca<br>
<a href="mailto:nia.hughes@oicr.on.ca">nia.hughes@oicr.on.ca</a>


Nia is the Program Manager for Bioinformatics.ca, where she coordinates the Canadian Bioinformatics Workshop Series. Prior to starting at the OICR, she completed her M.Sc in Bioinformatics from the University of Guelph in 2020 before working there as a bioinformatician studying epigenetic and transcriptomic patterns across maize varieties.

<p></p>

## Pre-workshop Materials and Laptop Setup Instructions {-}

### Laptop Setup Instructions {-}

A checklist to setup your laptop can be found here. 

Install these tools on your laptop before coming to the workshop:

1. R (4.0+)
<i>Note:</i> MacBook users with an Apple silicon chip (e.g., M1 or M2) should install the "arm64" version of R, while MacBook users with an Intel chip should install the regular (64-bit) version of R. You can check your laptop's hardware specifications by clicking the Apple icon (top left corner) > About This Mac and verifying whether the chip is Apple or Intel.
2. Rstudio
3.  Make sure you have a robust internet browser such as Firefox, Safari or Chrome (not Internet Explorer).
4.  Make sure you have a PDF viewer (e.g. Adobe Acrobat, Preview or similar) or that you can read PDF files in your Web browser.

### R packages {-}

```{r install-packages, eval=TRUE}
pkgList <- c("tidyverse", "clValid","rgl","RColorBrewer","corrplot","ClusterR",
  "Rtsne","umap","BiocManager","mlbench","plotrix", "factoextra","bladderbatch",
  "airway")
for (cur in pkgList){
  message(sprintf("\tChecking for %s ...", cur))
  if (!requireNamespace(cur, quietly = TRUE)) install.packages(cur)
}
if (!requireNamespace("edgeR", quietly = TRUE)) BiocManager::install("edgeR")
```

We are going to use `titanic.csv` in Module 3.
Download [the dataset here](https://drive.google.com/file/d/1PdP7bZTbmcRrPPPYFf1Mr8vfucom-d4k/view?usp=sharing) and copy it into your working directory.

# Lecture slides {-}

* [Module 1: Exploratory Data Analysis and Clustering](https://github.com/RealPaiLab/CBW_AUR2023/blob/main/book/lectures/AUR_2023_Module1_EDA.pdf)
* [Module 2: Dimensionality reduction for visualization and analysis](https://github.com/RealPaiLab/CBW_AUR2023/blob/main/book/lectures/AUR_2023_Module2_DimensionReduction.pdf)
* [Module 3: Generalized linear models](https://github.com/RealPaiLab/CBW_AUR2023/blob/main/book/lectures/AUR_2023_Module3_GLM.pdf)
* [Module 4: Multiple hypothesis testing with RNA-seq differential expression analysis](https://github.com/RealPaiLab/CBW_AUR2023/blob/main/book/lectures/AUR_2023_Module4_RNAseq.pdf)


<!--chapter:end:index.Rmd-->

# Module 1: Exploratory Data Analysis and Clustering {-}

The goal of this lab is to examine correlation structure in a sample transcriptomic dataset using clustering.

We will start using a small mouse expression dataset from two tissues:

* We will briefly explore the dataset using `dim()`, `head()`, and `summary()`
* We will visualize correlation using `pairs()` and `corrplot()`
* We will cluster samples using k-means and hierarchical clustering methods using `dist()` and `hclust()`
  * We will assign samples to clusters using `cutree()` and visualize these using `heatmap()`
* We will use `clValid` to find out what cluster number best separates groups

Finally we will explore clustering in a more complex bladder cancer dataset. 

## Load `mouse` data {-}

For this exercise we are going to load a dataset built into the `clValid` package.
This dataset measures 147 genes and expressed sequence tags in two developing mouse lineages: the neural crest cells and mesoderm-derived cells.
There are three samples per group.

Let's load the data.

```{r, class.source="codeblock",eval=TRUE}
suppressMessages(library(clValid))
data(mouse)
```

Use `str()` to see what types of data the columns have:

```{r, class.source="codeblock",eval=TRUE}
str(mouse)
```

Another command is `head()`:
```{r, class.source="codeblock",eval=TRUE}
head(mouse)
```

Summary provides useful information about the distribution of variables. Note that `FC` has categorical variables:

```{r, class.source="codeblock",eval=TRUE}
summary(mouse)
```

What are the values in `FC`?

```{r, class.source="codeblock",eval=TRUE}
table(mouse$FC)
```

Usually the information about samples ("metadata") is in a different table. 

Let's load the sample information about the mouse dataset:
```{r,eval=TRUE}
pheno <- read.delim("data/mouse_pheno.txt",sep="\t",h=T,as.is=T)
```

Let's look at it:
```{r,eval=TRUE}
head(pheno)
```

Let's use `pairs()`  to look at pairwise scatterplots of the expression data in a single plot. We need to subset the columns with the expression data first:

```{r, class.source="codeblock",eval=TRUE}
mouse_exp = mouse[,c("M1","M2","M3","NC1","NC2","NC3")]
pairs(mouse_exp)
```

## Correlations, distances, and clustering {-}

Let's look at sample correlation matrix. This should be a square matrix, equal to the number of samples.
We have six samples, so the correlation matrix should be `6x6`.

```{r, class.source="codeblock",eval=TRUE}
library(corrplot)
mouse_cor <- cor(mouse_exp)
dim(mouse_cor)
round(mouse_cor,2)
corrplot(mouse_cor, method="color")
```

* Which samples appear to be best correlated with each other?
* Which samples don't appear to be as well correlated with each other?

## Hierarchical clustering {-}

Hierarchical clustering requires distances between samples. Let's use `dist()` to compute these distances, and `hclust()` to generate the hierarchical clustering object. 

```{r, class.source="codeblock",eval=TRUE}
d <- dist(t(log(mouse_exp)))
```
 
Using these distances, we cluster the samples using hierarchical clustering:

```{r,eval=TRUE}
h <- hclust(d,method="ward.D2")
```

The output of this can be plotted:

```{r,eval=TRUE}
plot(h)
```

Can you guess how many clusters could best fit the data?


Now let's add a heatmap to this dendrogram, so we can see the values of genes in each cluster. For this we will use the `heatmap()` function.
First let's just try the heatmap function:

```{r,eval=TRUE}
mouse_exp <- as.matrix(mouse_exp)
heatmap(mouse_exp)
```

We can clearly see the data separated into two. But now let's colour-code samples based on cluster assignments.
We get cluster assignments by "cutting" the dendrogram for two clusters (something we expect from our experimental design). 
We use `cutree()` for this.  
```{r, class.source="codeblock",eval=TRUE}
h2 <- cutree(h, k = 2)
```

Let's look at our assigned labels:
```{r,eval=TRUE}
h2
```

Let's assign colours to the clusters so that cluster 1 is in pink, and cluster 2 is in green:
```{r,eval=TRUE}
clust_colours <- c("pink","green")[h2]
```

Look at clust_colours:
```{r,eval=TRUE}
clust_colours
```

Now let's plot the heatmap using these assigned cluster labels:

```{r,eval=TRUE}
heatmap(mouse_exp,
    ColSideColors=clust_colours)
```

Note that the two colours are completely divided (i.e., there is no interspersed pink and green). 

However, not all datasets are this simple. Let's cluster a bladder cancer gene expression dataset.
This is in the R package, `bladderbatch` which we have already installed.

```{r,eval=TRUE}
library(bladderbatch)
```

Load the dataset:
```{r,eval=TRUE}
data(bladderdata)
```

We will use specialized functions to get the expression data and sample information. 
```{r,eval=TRUE}
bexprs <- exprs(bladderEset)
bpheno <- pData(bladderEset)
```

How many genes and samples do we have in this dataset?

Let us use the same code as above to cluster these samples:
```{r bladder-hclust,eval=TRUE}
d <- dist(t(bexprs))
h <- hclust(d, method="ward.D2")
plot(h)
```

How many clusters do we see?

Let's assume three clusters, and assign colours to these. As before we use `cutree()`:
```{r bladder-cutree,eval=TRUE}
h3 <- cutree(h, k=3)
clust_colours <- c("red","green","blue")[h3]
```

Look at the colour assignments?
```{r,eval=TRUE}
table(h3)
```

Let's just plot the heatmap.
```{r heatmap-bladder1,eval=TRUE}
heatmap(bexprs,
    ColSideColors=clust_colours)
```

Why aren't the samples clustering?

Now try providing the `hclustfun`  to `heatmap()` so it uses the same method to cluster as we did.
For this we will create a custom function:
```{r hclust-func,eval=TRUE}
myhclust <- function(x) {
    hclust(x,method="ward.D2")
}
```

And now we run heatmap again, using our clustering function:

```{r heatmap-bladder2,eval=TRUE}
heatmap(bexprs,
    ColSideColors=clust_colours,
    hclustfun=myhclust
)
```


## K-means clustering {-}

Let's try using k-means clustering, asking for three clusters:

```{r, class.source="codeblock",eval=TRUE}
kclust <- kmeans(
    mouse_exp, 
    centers = 3
)
kclust
```

## Using `clValid` to determine number of clusters {-}

Use the `clValid()` function to validate clusters using the:

* Dunn index,
* silhouette scores, and
* connectivity

```{r, class.source="codeblock",eval=TRUE}
validation_data <- clValid(
    mouse_exp,
    2:6, # num. clusters to evaluate
    clMethods = c("hier","kmeans"), # methods to eval.
    validation = "internal"
)
```

Let's look at the results:
```{r, class.source="codeblock",eval=TRUE}
summary(validation_data)
```

All measures of clustering consistently indicate that **two** clusters best fit the data.

Now let's cluster:

```{r, class.source="codeblock",eval=TRUE}
d <- dist(t(log(mouse_exp)))
h <- hclust(d,method="ward.D2")
cluster_ids <- cutree(h, k = 2)
clust_colors <- c("dodgerblue","orangered")[cluster_ids]

heatmap(
    as.matrix(mouse_exp),
    hclustfun = myhclust,
    ColSideColors = clust_colors
)
```

<!--chapter:end:01-Module1_EDAClustering.Rmd-->

# Module 2: Dimensionality reduction {-}

The goal of this lab is to learn how to reduce the dimension of your dataset.

We will learn three different methods commonly used for dimension reduction:

1. Principal Component Analysis
2. t-stochastic Neighbour Embedding (tSNE)
3. Uniform Manifold Approximation (UMAP)


## Principal Component Analysis {-}

Let's start with PCA. PCA is commonly used as one step in a series of analyses. The goal of PCA is to explain most of the variability in the data with a smaller number of variables than the original data set. You can use PCA to explain the variability in your data using fewer variables. Typically, it is useful to identify outliers and determine if there's batch effect in your data.

Data: We will use the dataset that we used for exploratory analysis in Module 1. Load the mouse data.

```{r, class.source="codeblock",eval=TRUE}
library(clValid)
data("mouse")
mouse_exp <- mouse[,c("M1","M2","M3","NC1","NC2","NC3")]
head(mouse_exp)
```


### Step 1. Preparing Our Data {-}

It is important to make sure that all the variables in your dataset are on the same scale to ensure they are comparable. So, let us check if that is the case with our dataset. To do that, we will first compute the means and variances of each variable using `apply()`.

```{r, class.source="codeblock",eval=TRUE}
apply(mouse_exp, 2, mean)
apply(mouse_exp, 2, var)
```

As you can see, the means and variances for all the six variables are almost the same and on the same scale, which is great!

However, keep in mind that, the variables need not always be on the same scale in other non-omics datasets. PCA is influenced by the magnitude of each variable. So, it is important to include a scaling step during data preparation. Ideally, it is great to have variables centered at zero for PCA because it makes comparing each principal component to the mean straightforward. Scaling can be done either using `scale()`. 

### Step 2. Apply PCA {-}

Since our variables are on the same scale, we can directly apply PCA using `prcomp()`.

```{r, class.source="codeblock",eval=TRUE}
pc_out <- prcomp(t(mouse_exp))
```

The output of `prcomp()` is a list. Examine the internal structure of `pc_out`.

```{r, class.source="codeblock",eval=TRUE}
str(pc_out)
```

The output of `prcomp()` contains five elements `sdev`, `rotation`, `center`, `scale` and `x`. Let us examine what each looks like. 

```{r, class.source="codeblock",eval=TRUE}
pc_out$sdev
```

`sddev` gives standard deviation (used for computing variance explained). We will see how in sections below.

```{r, class.source="codeblock",eval=TRUE}
head(pc_out$rotation)
```

After PCA, the observations are expressed in new axes and the loadings are provided in `pc_out$rotation`. Each column of `pc_out$rotation` contains the corresponding principal component loading vector.

We see that there are six distinct principal components, as indicated by column names of `pc_out$rotation`. 
```{r, class.source="codeblock",eval=TRUE}
pc_out$center
pc_out$scale
```

The `center` and `scale` elements correspond to the means and standard deviations of the variables that were used for scaling prior to implementing PCA.

```{r, class.source="codeblock",eval=TRUE}
# See the principal components
dim(pc_out$x)
head(pc_out$x)
```


Let’s now see the summary of the analysis using the `summary()` function!

```{r, class.source="codeblock",eval=TRUE}
summary(pc_out)
```

The first row gives the `Standard deviation` of each component, which is the same as the result of `pc_out$sdev`. 

The second row, `Proportion of Variance`, shows the percentage of explained variance, also obtained as variance/sum(variance) where variance is the square of sdev.

Compute variance
```{r, class.source="codeblock",eval=TRUE}
pc_out$sdev^2 / sum(pc_out$sdev^2)
```

From the second row you can see that the first principal component explains over 82.4% of the total variance (Note: multiply each number by 100 to get the percentages). 

The second principal component explains 8.2% of the variance, and the amount of variance explained reduces further down with each component. 

Finally, the last row, `Cumulative Proportion`, calculates the cumulative sum of the second row. 

Now, let's have some fun with visualising the results of PCA.

### Step 3. Visualisation of PCA results {-}

**A. Scree plot**

We can visualize the percentage of explained variance per principal component by using what is called a scree plot. We will call the `fviz_eig()` function of the factoextra package for the application. You may need to install the package using `install.packages("factoextra")`.

```{r, class.source="codeblock",eval=TRUE}
library(factoextra)
fviz_eig(pc_out, 
         addlabels = TRUE)
```

The x-axis shows the PCs and the y-axis shows the percentage of variance explained that we saw above. Percentages are listed on top of the bars. It’s common to see that the first few principal components explain the major amount of variance.

Scree plot can also be used to decide the number of components to keep for rest of your analysis. One of the ways is using the elbow rule. This method is about looking for the “elbow” shape on the curve and retaining all components before the point where the curve flattens out. Here, the elbow appears to occur at the second principal component. 

Note that we will NOT remove any components for the current analysis since our goal is to understand how PCA can be used to identify batch effect in the data.

**B. Scatter plot**

After a PCA, the observations are expressed in principal component scores (as we saw above in `pc_out$rotation`). So, it is important to visualize the observations along the new axes (principal components) how observations have been transformed and to understand the relations in the dataset.

This can be achieved by drawing a scatterplot. To do so, first, we need to extract and the principal component scores in `pc_out$rotation`, and then we will store them in a data frame called PC.

```{r, class.source="codeblock",eval=TRUE}
PC <- as.data.frame(pc_out$x)
```

Plot the first two principal components as follows:

```{r, class.source="codeblock",eval=TRUE, fig.width=7, fig.height=7}
plot(x = PC$PC1,
     y = PC$PC2,
     pch = 19,
     xlab="PC1",
     ylab="PC2")
```

We see six points in two different groups. The points correspond to six samples. But, we don't know what group/condition they belong to.

That can be done by adding sample-related information to the data.frame `PC` (such as cell type, treatment type, batch they were processed etc) as new variables. Here we will add the sample names and the cell types. 

```{r, class.source="codeblock",eval=TRUE}
PC$sample <- factor(rownames(PC))
PC$cells <- factor(c(rep("mesoderm", 3), rep("neural_crest", 3)))
```

Plot the scatterplot again and now, colour the points by cell type. Then, add sample names and legend.

```{r, , class.source="codeblock",eval=TRUE, fig.width=7, fig.height=7}
plot(x = PC$PC1,
     y = PC$PC2,
     col = PC$cells,
     pch = 19,
     xlab="PC1",
     ylab="PC2")

text(x= PC$PC1,
     y = PC$PC2-0.15,
     labels = PC$sample)

legend("bottomright", 
       legend = levels(PC$cells), 
       col = seq_along(levels(PC$cells)), 
       pch = 19)
 
```

Samples from each cell type are closer together on the scatter plot. If the batch information is available, it can also be used to colour the scatterplot. Ideally, samples from different conditions should cluster together, irrespective of the batch they were processed in.

You can also plot other PCs such as PC2 vs PC3 by changing the x and y variables above. Another way to plot all PCs is using `pairs()`
```{r, class.source="codeblock",eval=TRUE}
pairs(PC[,1:6], 
      col=c("black","red")[PC$cells], 
      pch=16)
```


**C. Biplot**

Another useful plots to understand the results are biplots. We will use the `fviz_pca_biplot()` function of the factoextra package. We will set label="var" argument to label the variables.

```{r, class.source="codeblock",eval=TRUE, fig.width=7, fig.height=7}
fviz_pca_biplot(pc_out, label = "var")
```

The axes show the principal component scores, and the vectors are the loading vectors, whose components are in the magnitudes of the loadings. Vectors indicate that samples from each cell type are closer together. 

## t-Distributed Stochastic Neighbor Embedding (t-SNE) {-}

t-SNE is a technique for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets.

There are several packages that have implemented t-SNE. Here we are going to use  the package `tsne` and the function `tsne`. Let’s run the t-SNE algorithm on the iris dataset and generate a t-SNE plot.

```{r, class.source="codeblock",eval=TRUE}
library(tsne)
library(RColorBrewer)

### load the input data
data(iris)
iris_data <- iris[,-5]

# set colours of the plot
my_cols_vec <- brewer.pal("Set1",n = length(levels(iris$Species)))
species_cols <- my_cols_vec[factor(iris$Species)]

# run t-SNE
iris_tsne <- tsne(iris_data)

plot(iris_tsne,
     pch=16, 
     col=species_cols)
legend("topright",
       legend = levels(iris$Species),
       col = my_cols_vec,
       pch = 19)
```

The `tsne` function has a parameter called `perplexity` which determines how to balance attention to neighborhood vs global structure. Default value is 30 which was used above. Set perplexity to 10, 20, 50, 100 and rerun tsne. Then visualise each result.

```{r, class.source="codeblock",eval=TRUE}
iris_tsne10 <- tsne(iris_data,perplexity = 10)
iris_tsne20 <- tsne(iris_data,perplexity = 20)
iris_tsne50 <- tsne(iris_data,perplexity = 50)
iris_tsne100 <- tsne(iris_data,perplexity = 100)
```


```{r, fig.width=7, fig.height=7, class.source="codeblock",eval=TRUE}
par(mfrow=c(2,2))

plot(iris_tsne10[,1],
     iris_tsne10[,2],
     main = "Perplexity = 10",
     col = species_cols,
     pch=16)

plot(iris_tsne20[,1],
     iris_tsne20[,2],
     main = "Perplexity = 20",
     col = species_cols,
     pch=16)

plot(iris_tsne50[,1],
     iris_tsne50[,2],
     main = "Perplexity = 50",
     col = species_cols,
     pch=16)

plot(iris_tsne100[,1],
     iris_tsne100[,2],
     main = "Perplexity = 100",
     col = species_cols,
     pch=16)
```

Higher perplexity leads to higher spread in your data.

## Uniform Manifold Approximation and Projection (UMAP) {-}

UMAP is another dimension reduction method and it uses similar neighborhood approach as t-SNE except uses Riemannian geometry.

Here we are going to use  the package `umap` and the function `umap`. Let’s apply UMAP on the iris dataset and generate a UMAP plot.

```{r, class.source="codeblock",eval=TRUE}
## umap
library(umap)

iris_umap <- umap(iris_data)
str(iris_umap)
```


```{r, class.source="codeblock",eval=TRUE}
par(mfrow=c(1,1))
plot(iris_umap$layout[,1],
     iris_umap$layout[,2],
     col = species_cols, pch = 19)
# legend("topright",
#         legend = levels(mouse$FC),
#         col = species_cols,
#         pch = 19)
```



<!--chapter:end:11-Module2_PCA.Rmd-->

# Module 2: Exercise Results {-}

PCA
```{r, class.source="codeblock",eval=FALSE}
c_pcs = prcomp(crabs_meas)
```

Plot PC projections (embeddings). 
```{r, class.source="codeblock",eval=FALSE}
pairs(c_pcs$x, col = c("orchid","forestgreen")[factor(crabs$sp)])
pairs(c_pcs$x, col = c("orchid","forestgreen")[factor(crabs$sex)])
```

tSNE:

```{r, class.source="codeblock",eval=FALSE}
library(tsne)
c_tsne10 = tsne(crabs_meas,perplexity = 10)
c_tsne20 = tsne(crabs_meas,perplexity = 20)
c_tsne50 = tsne(crabs_meas,perplexity = 50)
c_tsne100 = tsne(crabs_meas,perplexity = 100)
```


sex_cols = c("orchid","forestgreen")[factor(crabs$sex)]

Color-code tSNE plot by species, try various perplexity levels:
```{r, class.source="codeblock",eval=FALSE}
species_cols = c("orchid","forestgreen")[factor(crabs$sp)]
par(mfrow=c(2,2))
plot(c_tsne10[,1],
     c_tsne10[,2],
     main = "Perplexity = 10",
     col = species_cols)

plot(c_tsne20[,1],
     c_tsne20[,2],
     main = "Perplexity = 20",
     col = species_cols)
plot(c_tsne50[,1],
     c_tsne50[,2],
     main = "Perplexity = 50",
     col = species_cols)
plot(c_tsne100[,1],
     c_tsne100[,2],
     main = "Perplexity = 100",
     col = species_cols)
```

Now do the same, but colour-code for sex:

```{r, class.source="codeblock",eval=FALSE}
sex_cols = c("orchid","forestgreen")[factor(crabs$sex)]
par(mfrow=c(2,2))
plot(c_tsne10[,1],
     c_tsne10[,2],
     main = "Perplexity = 10",
     col = sex_cols)
plot(c_tsne20[,1],
     c_tsne20[,2],
     main = "Perplexity = 20",
     col = sex_cols)
plot(c_tsne50[,1],
     c_tsne50[,2],
     main = "Perplexity = 50",
     col = sex_cols)
plot(c_tsne100[,1],
     c_tsne100[,2],
     main = "Perplexity = 100",
     col = sex_cols)
```


Run UMAP
```{r, class.source="codeblock",eval=FALSE}
library(umap)
c_umap <- umap(crabs_meas)
str(c_umap)

par(mfrow=c(1,2))
plot(c_umap$layout[,1],
     c_umap$layout[,2],
     col = species_cols, pch = 19, 
     main = "Colored by species")

plot(c_umap$layout[,1],
     c_umap$layout[,2],
     col = sex_cols, pch = 19, 
     main = "Colored by sex")
```

<!--chapter:end:12-Module2_Exercise.Rmd-->

# Module 3: Generalized Linear Models {-}

 In this module we're going to cover:

 * Reading data from files and evaluating missingness
 * Creating publication-quality plots with `ggplot2`
 * Fitting models with binary outcomes using generalized linear models

## Essential R: Reading tables from files, merging, basic data exploration  {-}

For this we're going to use two data files available in the course data directory.

[Download these](https://drive.google.com/drive/folders/1hKRa7ui-DmQCbqoKlgkmvhlrbXDmbgsV?usp=sharing) for use and put them in your R working directory.

* `tomerge1.csv` : comma-separated values
* `tomerge2.txt` : space-delimited values

Use `read.delim()` to read in tables. Note the use of the `sep=` parameter to indicate what the column separator is:

```{r, class.source="codeblock",eval=TRUE}
x1 <- read.delim("data/tomerge1.csv",sep=",")
head(x1)

x2 <- read.delim("data/tomerge2.txt",sep=" ")
head(x2)
```

Use `merge()` to combine the two tables by sample ID. Note the use of `by.x` and `by.y` to tell `merge()` which columns are equivalent:

```{r, class.source="codeblock", eval=TRUE}
x_merge <- merge(x1, x2, by.x = "Sample_ID", by.y = "sampleID")
head(x_merge)
```

We're going to use a popular dataset for data analysis, pertaining to survival of passengers aboard the Titanic. 

Download [the dataset here](https://drive.google.com/file/d/1PdP7bZTbmcRrPPPYFf1Mr8vfucom-d4k/view?usp=sharing) and copy it into your working directory.

Let's read in the data from a file:
```{r, class.source="codeblock",eval=TRUE}
dat <- read.delim(
    "data/titanic.csv",
    sep="," # indicate the column separator
    ) 
```

Examine the columns:
```{r, class.source="codeblock",eval=TRUE}
head(dat)
```

Some of the columns are categorical, use `table()` to look at the tallies:
Examine the columns:
```{r, class.source="codeblock",eval=TRUE}
table(dat$Survived)
table(dat$Pclass)
```

Use `summary()` to look at continuous-valued data:
```{r, class.source="codeblock",eval=TRUE}
summary(dat$Age)
```

Notice that there are 177 `NA` (missing) values for age.
Let's visualize the missing data more systematically.

## Explore missing data {-}

For this let's use a little script that converts a table into black and white squares to visualize missing data. For this, install the `plotrix` package

```{r, class.source="codeblock",eval=TRUE}
if (!requireNamespace("plotrix", quietly = TRUE)) install.packages("plotrix")

suppressMessages(require(plotrix))

#' show data missingness as a chequered matrix
#' 
#' @param x (matrix) data matrix.
#' @param outFile (char) path to file for printing graph
#' @param wd (numeric) width in inches
#' @param ht (numeric) height in inches
#' @return plots missingness matrix to file
#' @import plotrix
#' @export
plotMissMat <- function(x,xlab="columns",
		ylab="rows",border=NA) {
	
	x <- !is.na(x)
	class(x) <- "numeric"
	color2D.matplot(x,show.values=FALSE,axes=FALSE,
		cs1=c(1,0),cs2=c(1,0),cs3=c(1,0),border=border,
		cex=0.8,
		xlab=xlab,ylab=ylab)
}
```

Let's look at the missingness in the Titanic dataset. Missing data is shown as a white cell, and non-missing data is shown in black.

```{r, class.source="codeblock",eval=TRUE}
plotMissMat(dat)
```

We can see a column with many missing values. This is probably the "age" data. 
Let's count the number of missing values on a per-column level.

For this we combine `is.na()`, which returns a TRUE/FALSE value for NA values, and `colSums()` which adds up the TRUE values down the columns.

```{r count-missing, class.source="codeblock",eval=TRUE}
colSums(is.na(dat))
```
This confirms that `Age` is the only column with missing data.

Now let's explore the data using plots.

## Essential R: Plots with `ggplot2`  {-}

`ggplot2` is a popular plotting package that uses an additive paradigm to build plots. 
 
Useful websites:
* [ggplot2 cheatsheet](https://rstudio.github.io/cheatsheets/html/data-visualization.html)
* The [ggplot2 website](https://ggplot2.tidyverse.org/) is a wealth of reference information, so here we just touch on the basics to get you started. 


Anytime you need to generate a specific kind of plot, the website will most likely have documentation for how to achieve it.

Let's start by creating a scatterplot with two **continuous variables**. For this let's load a dataset measuring statistics around quality of life in US states in the late 70's:

```{r, class.source="codeblock",eval=TRUE}
state.x77 <- as.data.frame(state.x77)
head(state.x77)
```

Create a base plot using the `ggplot` function. Then "add" a scatterplot to it. Notice that the plot has been assigned to a variable named `p`. 

This setup is standard for `ggplot2` and allows multiple visualizations to be applied to the same base plot.

We use `aes` to tell ggplot what the `x` and `y` axes are, and later, if we want to colour-code by a particular column.

```{r load-states-data, class.source="codeblock",eval=TRUE}
library(ggplot2)
p <- ggplot(state.x77, 
    aes(x = Illiteracy,y = Income)
    )

p <- p + geom_point() # scatter plot
p
```

Now let's add confidence intervals:
```{r states-geomsmooth, class.source="codeblock",eval=TRUE}
p + geom_smooth()
```

It looks like there is a negative relationship between illiteracy and income.
We can confirm this by looking at correlation:
```{r test-cor, class.source="codeblock",eval=TRUE}
x <- state.x77$Illiteracy
y <- state.x77$Income
cor.test(x,y)
cor.test(x,y)$p.value
```

Let's now examine **categorical variables**. 
In the titanic set, let's look at the fare paid based on passenger class:
```{r boxplot, class.source="codeblock",eval=TRUE}
p <- ggplot(dat)
p + geom_boxplot(
    aes(x = factor(Pclass), # "factor()" makes a data column a categorical variable
        y = Fare))
```

We can use barplots to examine counts and proportions.
Let's look at number of survivors, split by passenger class.

Here the `fill` command is used to to split each barplot by the category, "Survived".
So you can see number of passengers by "Pclass" split by survival.

```{r  barplot, class.source="codeblock",eval=TRUE}
p + geom_bar(
    aes(fill=factor(Survived), 
        Pclass)
)
```

The plot above shows count data. Let's convert this to proportions. We can see that the fraction of non-survivors in "Class 3" is high.
```{r barplot2 , class.source="codeblock",eval=TRUE}
p + geom_bar(
    aes(fill=factor(Survived), Pclass),
    position = "fill"
)
```

How about males versus females?
```{r finish-titanic, class.source="codeblock",eval=TRUE}
p + geom_bar(
    aes(fill=factor(Survived), Sex),
    position = "fill"
)
```

**Exercise for later:** Try other `ggplot` functions on these data or on a small data table of your project.
(avoid using large genomics datasets, because those are going to be harder to interpret)

## Fit binary response variable using `glm()` and logistic regression {-}

Let's fit a model to a binary outcome. For this we load a dataset that measures physiological variables in a cohort of Pima Indians.

```{r load-pima, class.source="codeblock",eval=TRUE}
library(mlbench)
data(PimaIndiansDiabetes2)
# type ?PimaIndiansDiabetes2 to learn more about the dataset.

dat <- PimaIndiansDiabetes2
head(dat)
```

Let's look at the impact of blood glucose levels on diabetes diagnosis. 

First let's make a scatterplot.  Could there be a relationship?

```{r plot-diabetes-vs-glucose, class.source="codeblock",eval=TRUE}
p <- ggplot(dat, aes(x = glucose, y = factor(diabetes)))
p + geom_point()
```

Could this be fit with a linear model? 

Is there a continuous line that could reasonably fit the data?

```{r, class.source="codeblock",eval=TRUE}
p <- ggplot(dat, aes(x = glucose, y = factor(diabetes)))
p + geom_point() + geom_smooth()
```

This is a situation where a logistic regression model would be an appropriate choice because of the binary outcome. 

We're going to use `glm()` to fit a model to these data:

```{r diabetes-model , class.source="codeblock",eval=TRUE}
mod <- glm(factor(diabetes)~glucose, 
    dat,
    family = "binomial" # set to model binary outcome
)
summary(mod)
```

Which factors explain a diabetes diagnosis?

What if we include a couple other factors?
```{r diabetes-model-expanded, class.source="codeblock",eval=TRUE}
mod <- glm(factor(diabetes)~ glucose + pregnant + age + pressure + triceps,
    dat,
    family = "binomial")
summary(mod)
```

```{block,type="rmd-caution"}
Note: This morning's session only intends to introduce you to fitting non-linear models. 

In practice you may need to do more work to test multiple models to ascertain best fits your data, using measures such as goodness-of-fit. You will also likely compute odds ratio (odds of increased Y per unit increase X), which is out of scope for the current tutorial. 

We strongly recommend that you learn these topics before applying these methods to your own data.
```

<!--chapter:end:21-Module3_LinearModels.Rmd-->

# Module 4: Finding differentially expressed genes from RNAseq data {-}

In this exercise, we will use `edgeR` to call differentially-expressed genes. For the example we will use an RNAseq dataset from a treatment-vehicle design. 

If you haven't already done so, install the `airway` dataset:

```{r,eval=TRUE}
# BiocManager::install("airway") 
```
These data are from the paper:

Himes BE, Jiang X, Wagner P, Hu R, Wang Q, Klanderman B, Whitaker RM, Duan Q, Lasky-Su J, Nikolos C, Jester W, Johnson M, Panettieri R Jr, Tantisira KG, Weiss ST, Lu Q. “RNA-Seq Transcriptome Profiling Identifies CRISPLD2 as a Glucocorticoid Responsive Gene that Modulates Cytokine Function in Airway Smooth Muscle Cells.” PLoS One. 2014 Jun 13;9(6):e99625. PMID: [24926665](https://pubmed.ncbi.nlm.nih.gov/24926665/).

From the abstract of the original paper: 
 “Using RNA-Seq, a high-throughput sequencing method, we characterized transcriptomic changes in four primary human ASM cell lines that were treated with dexamethasone - a potent synthetic glucocorticoid (1 micromolar for 18 hours).”

Let's load the data
```{r, eval=TRUE}
suppressMessages(library(airway))
data(airway)
```

These data happen to be in a Bioconductor-specific format, so we use the special functions, `assay()` and `colData()` to get the expression data and smaple information. 

```{r, eval=TRUE}
xpr <- assay(airway)
samples <- colData(airway)
```

**Questions:**  
* How many samples are in this experiment?
* How many genes were measured?
* How many treatment groups are there (`dex` column)?


Let's created a `DGEList` object (DGE stands for "Differential Gene Expression"). This object is what we will use for our differential expression analysis.

Note: Make phenotype of interest categorical. In R that means converting to a `factor` type with categorical `levels`. You can think of levels as ordinal representations (e.g., first level = 1, second = 2, etc., )

If `levels=` are not set, the default uses alphabetical order.  We recommend explicitly setting levels so that there are no assumptions.

Load the `edgeR` package:

```{r , class.source="codeblock",eval=TRUE}
suppressMessages(require(edgeR))
```

Let's create a `DGEList` object for the differential expression analysis.  Note that `group` must be a categorical variable (use `factor()` to convert it to one):

```{r dge-create, class.source="codeblock",eval=TRUE}
group <- factor(samples$dex)
dge <- DGEList(
    counts = xpr,
    group = group
    )
```

Remove low-count genes: To filter low count genes, we're going to use a normalized count measure called `cpm` (counts per million). We are going to keep genes with 100 or greater counts per million for at least two samples:

```{r dge-seecounts, class.source="codeblock",eval=TRUE}
head(dge$counts)
```

Look at counts per million using `cpm`:

```{r dge-cpm, class.source="codeblock",eval=TRUE}
cpm(dge)[1:5,1:5]
```
This next line is a bit complex so let's unpack it:

*  We are using `cpm(dge)>100` as a logical test ("which genes have cpm > 100?"). 
*  For each gene, we want that test to be true for at least two samples. For this we use `rowSums()` to add up how many samples meet that criteria. 
  
```{r dge-libsize, class.source="codeblock",eval=TRUE}
dim(dge) #before 

# keep genes which have cpm>100 in 2 or more samples
tokeep <- rowSums(cpm(dge)>100) >= 2 

# now filter for these
dge <- dge[tokeep,keep.lib.sizes = FALSE]

# how many genes do we have left?
dim(dge) #after
```

Normalize the data:
```{r dge-calcnorm, class.source="codeblock",eval=TRUE}
dge <- calcNormFactors(dge)
```

Visualize the data:
```{r, class.source="codeblock",eval=TRUE}
plotMDS(
    dge, 
    col=as.numeric(dge$samples$group), 
    pch=16
)
legend(
    "bottomleft", 
    as.character(unique(dge$samples$group)),
    col=c(1,2), pch=16
    )
```

Let's create a model design to identify genes with a `group` effect:

```{r dge-model, class.source="codeblock",eval=TRUE}
group <- dge$samples$group
mod <- model.matrix(~group)
```

Estimate variation ("dispersion") for each gene:
```{r dge-disp, class.source="codeblock",eval=TRUE}
dge <- estimateDisp(dge, mod)
```

Call differentially expressed genes. 

Here we:

* fit a model for each gene, using `glmFit`
* we have built in an estimate of gene-wise dispersion to better identify treatment effect (or "contrast")
* for each gene, we run a likelihood ratio test which compares which model fits the data better: a null model (treatment effect = 0)
 or a full model (treatment effect is non-zero)

Note that `coef=2` fetches the effects for the treatment effect; `coef=1` would fetch effects of the intercept term.<br>
&nbsp; <br>
<img src="images/model_coefs.png" width="75%">


```{r dge-diffex, class.source="codeblock",eval=TRUE}
fit <- glmFit(dge,mod)
diffEx <- glmLRT(fit, coef = 2) # get coefficients for group term
```

Look at the top 10 differentially expressed genes:
```{r dge-toptags, class.source="codeblock",eval=TRUE}
tt <- topTags(diffEx, n=10)
tt
```

For the next steps we're going to need stats on all the genes we've tested. So let's get those:
```{r, top-tags-full,class.source="codeblock",eval=TRUE}
tt <- as.data.frame(
    topTags(diffEx, n=nrow(dge)
    )
)
```

A QQplot directly compares the pvalues from our statistical tests to the expected values from a random uniform distribution (p-value selected at random).

A deviation from the x=y line (diagonal) towards the top indicates an enrichment of signal.

```{r, qqplot, class.source="codeblock",eval=TRUE}
qqplot(
    tt$PValue, 
    runif(nrow(tt)), # randomly sample from uniform distribution
    xlab="p-values from real data",
    ylab="Randomly-sampled values from Uniform distribution",
    pch=16,cex=0.5
)

# x=y line as reference
abline(0,1,col="red")
```

Now let's call differentially expressed genes using the `decideTestDGE()` function and use `summary()` to see how many genes are upregulated (value `+1`), downregulated (value `-1`) and not called as changed (value `0`)
```{r, decide-tests-dge, class.source="codeblock",eval=TRUE}
diffEx2 <- decideTestsDGE(diffEx, 
    adjust.method="BH", 
    p.value=0.05
)
summary(diffEx2)
```

A **volcano plot** can help visualize effect magnitude - log2 fold-change or `log2FC` in the table ` against the corresponding p-value. Here we create a volcano plot, and colour-code upregulated genes in red, and downregulated genes in blue.

Note that we are combining two different tables, `tt` and `diffEx2` so we need to ensure the order is the same. Otherwise the colours will be in the wrong order *(try it for yourself!)*.

```{r, class.source="codeblock",eval=TRUE}
midx <- match(rownames(tt), rownames(diffEx2))
diffEx2 <- diffEx2[midx,]

cols <- rep("black",nrow(diffEx2))
cols[which(diffEx2>0 )]  <- "red"
cols[which(diffEx2<0)]  <- "blue"
# volcano plot
plot(tt$logFC,-log10(tt$PValue),pch=16,
    col=cols)
abline(v=0,lty=3)
```

Finally we can write our differential expression results out to file:
```{r, class.source="codeblock",eval=TRUE}
write.table(tt,file="diffEx.results.txt",
    sep="\t",
    col=TRUE,
    row=TRUE,
    quote=FALSE
)
```

<!--chapter:end:31-Module4_RNAseq.Rmd-->


<!--chapter:end:41-references.Rmd-->

