<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 1: Exploratory Data Analysis and Clustering | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 1: Exploratory Data Analysis and Clustering | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 1: Exploratory Data Analysis and Clustering | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="lecture-slides.html"/>
<link rel="next" href="module-2-dimensionality-reduction.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2024 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-1.-preparing-our-data"><i class="fa fa-check"></i>Step 1. Preparing Our Data</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-2.-apply-pca"><i class="fa fa-check"></i>Step 2. Apply PCA</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-3.-visualisation-of-pca-results"><i class="fa fa-check"></i>Step 3. Visualisation of PCA results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#exercise"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-exercise-results.html"><a href="module-2-exercise-results.html"><i class="fa fa-check"></i>Module 2: Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes from RNAseq data</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-1-exploratory-data-analysis-and-clustering" class="section level1 unnumbered hasAnchor">
<h1>Module 1: Exploratory Data Analysis and Clustering<a href="module-1-exploratory-data-analysis-and-clustering.html#module-1-exploratory-data-analysis-and-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The goal of this lab is to examine correlation structure in a sample transcriptomic dataset using clustering.</p>
<p>We will start using a small mouse expression dataset from two tissues:</p>
<ul>
<li>We will briefly explore the dataset using <code>dim()</code>, <code>head()</code>, and <code>summary()</code></li>
<li>We will visualize correlation using <code>pairs()</code> and <code>corrplot()</code></li>
<li>We will cluster samples using k-means and hierarchical clustering methods using <code>dist()</code> and <code>hclust()</code>
<ul>
<li>We will assign samples to clusters using <code>cutree()</code> and visualize these using <code>heatmap()</code></li>
</ul></li>
<li>We will use <code>clValid</code> to find out what cluster number best separates groups</li>
</ul>
<p>Finally we will explore clustering in a more complex bladder cancer dataset.</p>
<div id="load-mouse-data" class="section level2 unnumbered hasAnchor">
<h2>Load <code>mouse</code> data<a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this exercise we are going to load a dataset built into the <code>clValid</code> package.
This dataset measures 147 genes and expressed sequence tags in two developing mouse lineages: the neural crest cells and mesoderm-derived cells.
There are three samples per group.</p>
<p>Let’s load the data.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb17-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb17-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(clValid))</span>
<span id="cb17-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb17-2" tabindex="-1"></a><span class="fu">data</span>(mouse)</span></code></pre></div>
<p>Use <code>str()</code> to see what types of data the columns have:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb18-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb18-1" tabindex="-1"></a><span class="fu">str</span>(mouse)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    147 obs. of  8 variables:
##  $ ID : Factor w/ 147 levels &quot;1415787_at&quot;,&quot;1415904_at&quot;,..: 111 88 93 74 138 103 46 114 112 24 ...
##  $ M1 : num  4.71 3.87 2.88 5.33 5.37 ...
##  $ M2 : num  4.53 4.05 3.38 5.5 4.55 ...
##  $ M3 : num  4.33 3.47 3.24 5.63 5.7 ...
##  $ NC1: num  5.57 5 3.88 6.8 6.41 ...
##  $ NC2: num  6.92 5.06 4.46 6.54 6.31 ...
##  $ NC3: num  7.35 5.18 4.85 6.62 6.2 ...
##  $ FC : Factor w/ 9 levels &quot;ECM/Receptors&quot;,..: 3 8 6 6 1 3 1 6 5 6 ...</code></pre>
<p>Another command is <code>head()</code>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb20-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb20-1" tabindex="-1"></a><span class="fu">head</span>(mouse)</span></code></pre></div>
<pre><code>##             ID       M1       M2       M3
## 1   1448995_at 4.706812 4.528291 4.325836
## 2 1436392_s_at 3.867962 4.052354 3.474651
## 3 1437434_a_at 2.875112 3.379619 3.239800
## 4   1428922_at 5.326943 5.498930 5.629814
## 5 1452671_s_at 5.370125 4.546810 5.704810
## 6   1448147_at 3.471347 4.129992 3.964431
##        NC1      NC2      NC3
## 1 5.568435 6.915079 7.353144
## 2 4.995836 5.056199 5.183585
## 3 3.877053 4.459629 4.850978
## 4 6.795194 6.535522 6.622577
## 5 6.407555 6.310487 6.195847
## 6 4.474737 5.185631 5.177967
##                       FC
## 1 Growth/Differentiation
## 2   Transcription factor
## 3          Miscellaneous
## 4          Miscellaneous
## 5          ECM/Receptors
## 6 Growth/Differentiation</code></pre>
<p>Summary provides useful information about the distribution of variables. Note that <code>FC</code> has categorical variables:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb22-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb22-1" tabindex="-1"></a><span class="fu">summary</span>(mouse)</span></code></pre></div>
<pre><code>##           ID            M1       
##  1415787_at:  1   Min.   :2.352  
##  1415904_at:  1   1st Qu.:4.188  
##  1415993_at:  1   Median :4.994  
##  1416164_at:  1   Mean   :5.166  
##  1416181_at:  1   3rd Qu.:6.147  
##  1416221_at:  1   Max.   :9.282  
##  (Other)   :141                  
##        M2              M3       
##  Min.   :2.139   Min.   :2.500  
##  1st Qu.:4.151   1st Qu.:4.207  
##  Median :5.043   Median :5.054  
##  Mean   :5.140   Mean   :5.231  
##  3rd Qu.:6.015   3rd Qu.:6.129  
##  Max.   :9.273   Max.   :9.228  
##                                 
##       NC1             NC2       
##  Min.   :2.100   Min.   :1.996  
##  1st Qu.:4.174   1st Qu.:4.136  
##  Median :4.996   Median :5.056  
##  Mean   :5.120   Mean   :5.134  
##  3rd Qu.:5.860   3rd Qu.:5.920  
##  Max.   :8.905   Max.   :8.954  
##                                 
##       NC3       
##  Min.   :2.125  
##  1st Qu.:4.293  
##  Median :4.974  
##  Mean   :5.118  
##  3rd Qu.:5.826  
##  Max.   :9.251  
##                 
##                       FC    
##  EST                   :31  
##  Transcription factor  :28  
##  Miscellaneous         :25  
##  ECM/Receptors         :16  
##  Growth/Differentiation:16  
##  Unknown               :10  
##  (Other)               :21</code></pre>
<p>What are the values in <code>FC</code>?</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb24-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb24-1" tabindex="-1"></a><span class="fu">table</span>(mouse<span class="sc">$</span>FC)</span></code></pre></div>
<pre><code>## 
##          ECM/Receptors 
##                     16 
##                    EST 
##                     31 
## Growth/Differentiation 
##                     16 
##   Kinases/Phosphatases 
##                      7 
##             Metabolism 
##                      8 
##          Miscellaneous 
##                     25 
##         Stress-induced 
##                      6 
##   Transcription factor 
##                     28 
##                Unknown 
##                     10</code></pre>
<p>Usually the information about samples (“metadata”) is in a different table.</p>
<p>Let’s load the sample information about the mouse dataset:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb26-1" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/mouse_pheno.txt&quot;</span>,<span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>,<span class="at">h=</span>T,<span class="at">as.is=</span>T)</span></code></pre></div>
<p>Let’s look at it:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb27-1" tabindex="-1"></a><span class="fu">head</span>(pheno)</span></code></pre></div>
<pre><code>##   SampleID         Type
## 1       M1  Mesenchymal
## 2       M2  Mesenchymal
## 3       M3  Mesenchymal
## 4      NC1 Neural-crest
## 5      NC2 Neural-crest
## 6      NC3 Neural-crest</code></pre>
<p>Let’s use <code>pairs()</code> to look at pairwise scatterplots of the expression data in a single plot. We need to subset the columns with the expression data first:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb29-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb29-1" tabindex="-1"></a>mouse_exp <span class="ot">=</span> mouse[,<span class="fu">c</span>(<span class="st">&quot;M1&quot;</span>,<span class="st">&quot;M2&quot;</span>,<span class="st">&quot;M3&quot;</span>,<span class="st">&quot;NC1&quot;</span>,<span class="st">&quot;NC2&quot;</span>,<span class="st">&quot;NC3&quot;</span>)]</span>
<span id="cb29-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb29-2" tabindex="-1"></a><span class="fu">pairs</span>(mouse_exp)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="correlations-distances-and-clustering" class="section level2 unnumbered hasAnchor">
<h2>Correlations, distances, and clustering<a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let’s look at sample correlation matrix. This should be a square matrix, equal to the number of samples.
We have six samples, so the correlation matrix should be <code>6x6</code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb30-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb30-1" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span></code></pre></div>
<pre><code>## corrplot 0.92 loaded</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb32-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb32-1" tabindex="-1"></a>mouse_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(mouse_exp)</span>
<span id="cb32-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb32-2" tabindex="-1"></a><span class="fu">dim</span>(mouse_cor)</span></code></pre></div>
<pre><code>## [1] 6 6</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb34-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb34-1" tabindex="-1"></a><span class="fu">round</span>(mouse_cor,<span class="dv">2</span>)</span></code></pre></div>
<pre><code>##       M1   M2   M3  NC1  NC2  NC3
## M1  1.00 0.98 0.98 0.84 0.81 0.78
## M2  0.98 1.00 0.95 0.84 0.81 0.79
## M3  0.98 0.95 1.00 0.82 0.78 0.75
## NC1 0.84 0.84 0.82 1.00 0.97 0.95
## NC2 0.81 0.81 0.78 0.97 1.00 0.99
## NC3 0.78 0.79 0.75 0.95 0.99 1.00</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb36-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb36-1" tabindex="-1"></a><span class="fu">corrplot</span>(mouse_cor, <span class="at">method=</span><span class="st">&quot;color&quot;</span>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<ul>
<li>Which samples appear to be best correlated with each other?</li>
<li>Which samples don’t appear to be as well correlated with each other?</li>
</ul>
</div>
<div id="hierarchical-clustering" class="section level2 unnumbered hasAnchor">
<h2>Hierarchical clustering<a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Hierarchical clustering requires distances between samples. Let’s use <code>dist()</code> to compute these distances, and <code>hclust()</code> to generate the hierarchical clustering object.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb37-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb37-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(<span class="fu">log</span>(mouse_exp)))</span></code></pre></div>
<p>Using these distances, we cluster the samples using hierarchical clustering:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb38-1" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d,<span class="at">method=</span><span class="st">&quot;ward.D2&quot;</span>)</span></code></pre></div>
<p>The output of this can be plotted:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb39-1" tabindex="-1"></a><span class="fu">plot</span>(h)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>Can you guess how many clusters could best fit the data?</p>
<p>Now let’s add a heatmap to this dendrogram, so we can see the values of genes in each cluster. For this we will use the <code>heatmap()</code> function.
First let’s just try the heatmap function:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb40-1" tabindex="-1"></a>mouse_exp <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(mouse_exp)</span>
<span id="cb40-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb40-2" tabindex="-1"></a><span class="fu">heatmap</span>(mouse_exp)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>We can clearly see the data separated into two. But now let’s colour-code samples based on cluster assignments.
We get cluster assignments by “cutting” the dendrogram for two clusters (something we expect from our experimental design).
We use <code>cutree()</code> for this.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb41-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb41-1" tabindex="-1"></a>h2 <span class="ot">&lt;-</span> <span class="fu">cutree</span>(h, <span class="at">k =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Let’s look at our assigned labels:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb42-1" tabindex="-1"></a>h2</span></code></pre></div>
<pre><code>##  M1  M2  M3 NC1 NC2 NC3 
##   1   1   1   2   2   2</code></pre>
<p>Let’s assign colours to the clusters so that cluster 1 is in pink, and cluster 2 is in green:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb44-1" tabindex="-1"></a>clust_colours <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;pink&quot;</span>,<span class="st">&quot;green&quot;</span>)[h2]</span></code></pre></div>
<p>Look at clust_colours:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb45-1" tabindex="-1"></a>clust_colours</span></code></pre></div>
<pre><code>## [1] &quot;pink&quot;  &quot;pink&quot;  &quot;pink&quot;  &quot;green&quot;
## [5] &quot;green&quot; &quot;green&quot;</code></pre>
<p>Now let’s plot the heatmap using these assigned cluster labels:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb47-1" tabindex="-1"></a><span class="fu">heatmap</span>(mouse_exp,</span>
<span id="cb47-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb47-2" tabindex="-1"></a>    <span class="at">ColSideColors=</span>clust_colours)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>Note that the two colours are completely divided (i.e., there is no interspersed pink and green).</p>
<p>However, not all datasets are this simple. Let’s cluster a bladder cancer gene expression dataset.
This is in the R package, <code>bladderbatch</code> which we have already installed.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb48-1" tabindex="-1"></a><span class="fu">library</span>(bladderbatch)</span></code></pre></div>
<p>Load the dataset:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb49-1" tabindex="-1"></a><span class="fu">data</span>(bladderdata)</span></code></pre></div>
<p>We will use specialized functions to get the expression data and sample information.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb50-1" tabindex="-1"></a>bexprs <span class="ot">&lt;-</span> <span class="fu">exprs</span>(bladderEset)</span>
<span id="cb50-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb50-2" tabindex="-1"></a>bpheno <span class="ot">&lt;-</span> <span class="fu">pData</span>(bladderEset)</span></code></pre></div>
<p>How many genes and samples do we have in this dataset?</p>
<p>Let us use the same code as above to cluster these samples:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb51-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(bexprs))</span>
<span id="cb51-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb51-2" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method=</span><span class="st">&quot;ward.D2&quot;</span>)</span>
<span id="cb51-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb51-3" tabindex="-1"></a><span class="fu">plot</span>(h)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/bladder-hclust-1.png" width="672" /></p>
<p>How many clusters do we see?</p>
<p>Let’s assume three clusters, and assign colours to these. As before we use <code>cutree()</code>:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb52-1" tabindex="-1"></a>h3 <span class="ot">&lt;-</span> <span class="fu">cutree</span>(h, <span class="at">k=</span><span class="dv">3</span>)</span>
<span id="cb52-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb52-2" tabindex="-1"></a>clust_colours <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;green&quot;</span>,<span class="st">&quot;blue&quot;</span>)[h3]</span></code></pre></div>
<p>Look at the colour assignments?</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb53-1" tabindex="-1"></a><span class="fu">table</span>(h3)</span></code></pre></div>
<pre><code>## h3
##  1  2  3 
## 16 27 14</code></pre>
<p>Let’s just plot the heatmap.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb55-1" tabindex="-1"></a><span class="fu">heatmap</span>(bexprs,</span>
<span id="cb55-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb55-2" tabindex="-1"></a>    <span class="at">ColSideColors=</span>clust_colours)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/heatmap-bladder1-1.png" width="672" /></p>
<p>Why aren’t the samples clustering?</p>
<p>Now try providing the <code>hclustfun</code> to <code>heatmap()</code> so it uses the same method to cluster as we did.
For this we will create a custom function:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb56-1" tabindex="-1"></a>myhclust <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb56-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb56-2" tabindex="-1"></a>    <span class="fu">hclust</span>(x,<span class="at">method=</span><span class="st">&quot;ward.D2&quot;</span>)</span>
<span id="cb56-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb56-3" tabindex="-1"></a>}</span></code></pre></div>
<p>And now we run heatmap again, using our clustering function:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb57-1" tabindex="-1"></a><span class="fu">heatmap</span>(bexprs,</span>
<span id="cb57-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb57-2" tabindex="-1"></a>    <span class="at">ColSideColors=</span>clust_colours,</span>
<span id="cb57-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb57-3" tabindex="-1"></a>    <span class="at">hclustfun=</span>myhclust</span>
<span id="cb57-4"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb57-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/heatmap-bladder2-1.png" width="672" /></p>
</div>
<div id="k-means-clustering" class="section level2 unnumbered hasAnchor">
<h2>K-means clustering<a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let’s try using k-means clustering, asking for three clusters:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb58-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb58-1" tabindex="-1"></a>kclust <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(</span>
<span id="cb58-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb58-2" tabindex="-1"></a>    mouse_exp, </span>
<span id="cb58-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb58-3" tabindex="-1"></a>    <span class="at">centers =</span> <span class="dv">3</span></span>
<span id="cb58-4"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb58-4" tabindex="-1"></a>)</span>
<span id="cb58-5"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb58-5" tabindex="-1"></a>kclust</span></code></pre></div>
<pre><code>## K-means clustering with 3 clusters of sizes 22, 64, 61
## 
## Cluster means:
##         M1       M2       M3      NC1
## 1 7.416536 7.406216 7.414799 7.548674
## 2 5.553148 5.499583 5.642404 5.426931
## 3 3.947440 3.946048 4.012209 3.922949
##        NC2      NC3
## 1 7.534414 7.520608
## 2 5.435363 5.353342
## 3 3.950984 4.004362
## 
## Clustering vector:
##   1   2   3   4   5   6   7   8   9  10 
##   2   3   3   2   2   3   2   2   2   3 
##  11  12  13  14  15  16  17  18  19  20 
##   1   3   2   1   3   2   1   1   3   2 
##  21  22  23  24  25  26  27  28  29  30 
##   2   2   3   3   3   3   1   2   3   2 
##  31  32  33  34  35  36  37  38  39  40 
##   2   1   3   2   2   3   3   3   2   3 
##  41  42  43  44  45  46  47  48  49  50 
##   2   1   1   3   3   1   3   3   3   3 
##  51  52  53  54  55  56  57  58  59  60 
##   2   3   1   1   2   2   3   3   2   3 
##  61  62  63  64  65  66  67  68  69  70 
##   3   1   2   3   2   3   2   1   2   3 
##  71  72  73  74  75  76  77  78  79  80 
##   2   2   2   2   2   1   3   3   2   1 
##  81  82  83  84  85  86  87  88  89  90 
##   2   1   3   2   2   3   1   2   3   2 
##  91  92  93  94  95  96  97  98  99 100 
##   3   3   3   3   3   3   3   2   3   2 
## 101 102 103 104 105 106 107 108 109 110 
##   3   2   3   3   3   3   1   2   3   2 
## 111 112 113 114 115 116 117 118 119 120 
##   2   3   2   2   2   2   1   1   3   2 
## 121 122 123 124 125 126 127 128 129 130 
##   2   2   1   3   2   2   2   3   2   2 
## 131 132 133 134 135 136 137 138 139 140 
##   2   2   2   3   3   3   3   3   3   2 
## 141 142 143 144 145 146 147 
##   2   3   2   1   2   2   2 
## 
## Within cluster sum of squares by cluster:
## [1]  81.44264 166.24343 193.59229
##  (between_SS / total_SS =  74.3 %)
## 
## Available components:
## 
## [1] &quot;cluster&quot;      &quot;centers&quot;     
## [3] &quot;totss&quot;        &quot;withinss&quot;    
## [5] &quot;tot.withinss&quot; &quot;betweenss&quot;   
## [7] &quot;size&quot;         &quot;iter&quot;        
## [9] &quot;ifault&quot;</code></pre>
</div>
<div id="using-clvalid-to-determine-number-of-clusters" class="section level2 unnumbered hasAnchor">
<h2>Using <code>clValid</code> to determine number of clusters<a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Use the <code>clValid()</code> function to validate clusters using the:</p>
<ul>
<li>Dunn index,</li>
<li>silhouette scores, and</li>
<li>connectivity</li>
</ul>
<div class="sourceCode" id="cb60"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb60-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-1" tabindex="-1"></a>validation_data <span class="ot">&lt;-</span> <span class="fu">clValid</span>(</span>
<span id="cb60-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-2" tabindex="-1"></a>    mouse_exp,</span>
<span id="cb60-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-3" tabindex="-1"></a>    <span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>, <span class="co"># num. clusters to evaluate</span></span>
<span id="cb60-4"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-4" tabindex="-1"></a>    <span class="at">clMethods =</span> <span class="fu">c</span>(<span class="st">&quot;hier&quot;</span>,<span class="st">&quot;kmeans&quot;</span>), <span class="co"># methods to eval.</span></span>
<span id="cb60-5"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-5" tabindex="-1"></a>    <span class="at">validation =</span> <span class="st">&quot;internal&quot;</span></span>
<span id="cb60-6"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb60-6" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s look at the results:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb61-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb61-1" tabindex="-1"></a><span class="fu">summary</span>(validation_data)</span></code></pre></div>
<pre><code>## 
## Clustering Methods:
##  hierarchical kmeans 
## 
## Cluster sizes:
##  2 3 4 5 6 
## 
## Validation Measures:
##                                  2       3       4       5       6
##                                                                   
## hierarchical Connectivity   5.3270 14.2528 20.7520 27.0726 30.6194
##              Dunn           0.1291  0.0788  0.0857  0.0899  0.0899
##              Silhouette     0.5133  0.4195  0.3700  0.3343  0.3233
## kmeans       Connectivity  13.2548 17.6651 37.3980 43.2655 50.6095
##              Dunn           0.0464  0.0873  0.0777  0.0815  0.0703
##              Silhouette     0.4571  0.4182  0.3615  0.3367  0.3207
## 
## Optimal Scores:
## 
##              Score  Method       Clusters
## Connectivity 5.3270 hierarchical 2       
## Dunn         0.1291 hierarchical 2       
## Silhouette   0.5133 hierarchical 2</code></pre>
<p>All measures of clustering consistently indicate that <strong>two</strong> clusters best fit the data.</p>
<p>Now let’s cluster:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb63-1"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-1" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(<span class="fu">log</span>(mouse_exp)))</span>
<span id="cb63-2"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-2" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d,<span class="at">method=</span><span class="st">&quot;ward.D2&quot;</span>)</span>
<span id="cb63-3"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-3" tabindex="-1"></a>cluster_ids <span class="ot">&lt;-</span> <span class="fu">cutree</span>(h, <span class="at">k =</span> <span class="dv">2</span>)</span>
<span id="cb63-4"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-4" tabindex="-1"></a>clust_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;dodgerblue&quot;</span>,<span class="st">&quot;orangered&quot;</span>)[cluster_ids]</span>
<span id="cb63-5"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-5" tabindex="-1"></a></span>
<span id="cb63-6"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-6" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb63-7"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-7" tabindex="-1"></a>    <span class="fu">as.matrix</span>(mouse_exp),</span>
<span id="cb63-8"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-8" tabindex="-1"></a>    <span class="at">hclustfun =</span> myhclust,</span>
<span id="cb63-9"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-9" tabindex="-1"></a>    <span class="at">ColSideColors =</span> clust_colors</span>
<span id="cb63-10"><a href="module-1-exploratory-data-analysis-and-clustering.html#cb63-10" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="lecture-slides.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-2-dimensionality-reduction.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
