<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3: Generalized Linear Models | Analysis Using R</title>
  <meta name="description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3: Generalized Linear Models | Analysis Using R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3: Generalized Linear Models | Analysis Using R" />
  
  <meta name="twitter:description" content="Principles of exploratory data analysis, RNAseq differential expression, and generalized linear models" />
  

<meta name="author" content="Instructors: Shraddha Pai, Chaitra Sarathy" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-2-exercise-results.html"/>
<link rel="next" href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<img src="images/bioinfologo_big.png">
<li><a href="./">2024 CBW Analysis Using R Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet your Faculty</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-workshop-materials-and-laptop-setup-instructions"><i class="fa fa-check"></i>Pre-workshop Materials and Laptop Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#laptop-setup-instructions"><i class="fa fa-check"></i>Laptop Setup Instructions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#r-packages"><i class="fa fa-check"></i>R packages</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="lecture-slides.html"><a href="lecture-slides.html"><i class="fa fa-check"></i>Lecture slides</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html"><i class="fa fa-check"></i>Module 1: Exploratory Data Analysis and Clustering</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#load-mouse-data"><i class="fa fa-check"></i>Load <code>mouse</code> data</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#correlations-distances-and-clustering"><i class="fa fa-check"></i>Correlations, distances, and clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#hierarchical-clustering"><i class="fa fa-check"></i>Hierarchical clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#k-means-clustering"><i class="fa fa-check"></i>K-means clustering</a></li>
<li class="chapter" data-level="" data-path="module-1-exploratory-data-analysis-and-clustering.html"><a href="module-1-exploratory-data-analysis-and-clustering.html#using-clvalid-to-determine-number-of-clusters"><i class="fa fa-check"></i>Using <code>clValid</code> to determine number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html"><i class="fa fa-check"></i>Module 2: Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#principal-component-analysis"><i class="fa fa-check"></i>Principal Component Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-1.-preparing-our-data"><i class="fa fa-check"></i>Step 1. Preparing Our Data</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-2.-apply-pca"><i class="fa fa-check"></i>Step 2. Apply PCA</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#step-3.-visualisation-of-pca-results"><i class="fa fa-check"></i>Step 3. Visualisation of PCA results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i>t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i>Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="" data-path="module-2-dimensionality-reduction.html"><a href="module-2-dimensionality-reduction.html#exercise"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-exercise-results.html"><a href="module-2-exercise-results.html"><i class="fa fa-check"></i>Module 2: Exercise Results</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html"><i class="fa fa-check"></i>Module 3: Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration"><i class="fa fa-check"></i>Essential R: Reading tables from files, merging, basic data exploration</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#explore-missing-data"><i class="fa fa-check"></i>Explore missing data</a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2"><i class="fa fa-check"></i>Essential R: Plots with <code>ggplot2</code></a></li>
<li class="chapter" data-level="" data-path="module-3-generalized-linear-models.html"><a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression"><i class="fa fa-check"></i>Fit binary response variable using <code>glm()</code> and logistic regression</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html"><i class="fa fa-check"></i>Module 4: Finding differentially expressed genes from RNAseq data</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis Using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3-generalized-linear-models" class="section level1 unnumbered hasAnchor">
<h1>Module 3: Generalized Linear Models<a href="module-3-generalized-linear-models.html#module-3-generalized-linear-models" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this module we’re going to cover:</p>
<ul>
<li>Reading data from files and evaluating missingness</li>
<li>Creating publication-quality plots with <code>ggplot2</code></li>
<li>Fitting models with binary outcomes using generalized linear models</li>
</ul>
<div id="essential-r-reading-tables-from-files-merging-basic-data-exploration" class="section level2 unnumbered hasAnchor">
<h2>Essential R: Reading tables from files, merging, basic data exploration<a href="module-3-generalized-linear-models.html#essential-r-reading-tables-from-files-merging-basic-data-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this we’re going to use two data files available in the course data directory.</p>
<p><a href="https://drive.google.com/drive/folders/1hKRa7ui-DmQCbqoKlgkmvhlrbXDmbgsV?usp=sharing">Download these</a> for use and put them in your R working directory.</p>
<ul>
<li><code>tomerge1.csv</code> : comma-separated values</li>
<li><code>tomerge2.txt</code> : space-delimited values</li>
</ul>
<p>Use <code>read.delim()</code> to read in tables. Note the use of the <code>sep=</code> parameter to indicate what the column separator is:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb168-1"><a href="module-3-generalized-linear-models.html#cb168-1" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/tomerge1.csv&quot;</span>,<span class="at">sep=</span><span class="st">&quot;,&quot;</span>)</span>
<span id="cb168-2"><a href="module-3-generalized-linear-models.html#cb168-2" tabindex="-1"></a><span class="fu">head</span>(x1)</span></code></pre></div>
<pre><code>##   Sample_ID Exposure Biomarker_value
## 1         A        0              35
## 2         B        0              22
## 3         C        0              91
## 4         D        0               3
## 5         E        1              56
## 6         F        1              37</code></pre>
<div class="sourceCode" id="cb170"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb170-1"><a href="module-3-generalized-linear-models.html#cb170-1" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/tomerge2.txt&quot;</span>,<span class="at">sep=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb170-2"><a href="module-3-generalized-linear-models.html#cb170-2" tabindex="-1"></a><span class="fu">head</span>(x2)</span></code></pre></div>
<pre><code>##   sampleID Exposure_level
## 1        A     0.65405517
## 2        B     0.67202852
## 3        C     0.88646372
## 4        D     0.28433256
## 5        E     0.04166839
## 6        F     0.45263534
##   Biomarker2_detected
## 1                   1
## 2                   0
## 3                   1
## 4                   0
## 5                   0
## 6                   1</code></pre>
<p>Use <code>merge()</code> to combine the two tables by sample ID. Note the use of <code>by.x</code> and <code>by.y</code> to tell <code>merge()</code> which columns are equivalent:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb172-1"><a href="module-3-generalized-linear-models.html#cb172-1" tabindex="-1"></a>x_merge <span class="ot">&lt;-</span> <span class="fu">merge</span>(x1, x2, <span class="at">by.x =</span> <span class="st">&quot;Sample_ID&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;sampleID&quot;</span>)</span>
<span id="cb172-2"><a href="module-3-generalized-linear-models.html#cb172-2" tabindex="-1"></a><span class="fu">head</span>(x_merge)</span></code></pre></div>
<pre><code>##   Sample_ID Exposure Biomarker_value
## 1         A        0              35
## 2         B        0              22
## 3         C        0              91
## 4         D        0               3
## 5         E        1              56
## 6         F        1              37
##   Exposure_level Biomarker2_detected
## 1     0.65405517                   1
## 2     0.67202852                   0
## 3     0.88646372                   1
## 4     0.28433256                   0
## 5     0.04166839                   0
## 6     0.45263534                   1</code></pre>
<p>We’re going to use a popular dataset for data analysis, pertaining to survival of passengers aboard the Titanic.</p>
<p>Download <a href="https://drive.google.com/file/d/1PdP7bZTbmcRrPPPYFf1Mr8vfucom-d4k/view?usp=sharing">the dataset here</a> and copy it into your working directory.</p>
<p>Let’s read in the data from a file:</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb174-1"><a href="module-3-generalized-linear-models.html#cb174-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(</span>
<span id="cb174-2"><a href="module-3-generalized-linear-models.html#cb174-2" tabindex="-1"></a>    <span class="st">&quot;data/titanic.csv&quot;</span>,</span>
<span id="cb174-3"><a href="module-3-generalized-linear-models.html#cb174-3" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">&quot;,&quot;</span> <span class="co"># indicate the column separator</span></span>
<span id="cb174-4"><a href="module-3-generalized-linear-models.html#cb174-4" tabindex="-1"></a>    ) </span></code></pre></div>
<p>Examine the columns:</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb175-1"><a href="module-3-generalized-linear-models.html#cb175-1" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   PassengerId Survived Pclass
## 1           1        0      3
## 2           2        1      1
## 3           3        1      3
## 4           4        1      1
## 5           5        0      3
## 6           6        0      3
##                                                  Name
## 1                             Braund, Mr. Owen Harris
## 2 Cumings, Mrs. John Bradley (Florence Briggs Thayer)
## 3                              Heikkinen, Miss. Laina
## 4        Futrelle, Mrs. Jacques Heath (Lily May Peel)
## 5                            Allen, Mr. William Henry
## 6                                    Moran, Mr. James
##      Sex Age SibSp Parch           Ticket
## 1   male  22     1     0        A/5 21171
## 2 female  38     1     0         PC 17599
## 3 female  26     0     0 STON/O2. 3101282
## 4 female  35     1     0           113803
## 5   male  35     0     0           373450
## 6   male  NA     0     0           330877
##      Fare Cabin Embarked
## 1  7.2500              S
## 2 71.2833   C85        C
## 3  7.9250              S
## 4 53.1000  C123        S
## 5  8.0500              S
## 6  8.4583              Q</code></pre>
<p>Some of the columns are categorical, use <code>table()</code> to look at the tallies:
Examine the columns:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb177-1"><a href="module-3-generalized-linear-models.html#cb177-1" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Survived)</span></code></pre></div>
<pre><code>## 
##   0   1 
## 549 342</code></pre>
<div class="sourceCode" id="cb179"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb179-1"><a href="module-3-generalized-linear-models.html#cb179-1" tabindex="-1"></a><span class="fu">table</span>(dat<span class="sc">$</span>Pclass)</span></code></pre></div>
<pre><code>## 
##   1   2   3 
## 216 184 491</code></pre>
<p>Use <code>summary()</code> to look at continuous-valued data:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb181-1"><a href="module-3-generalized-linear-models.html#cb181-1" tabindex="-1"></a><span class="fu">summary</span>(dat<span class="sc">$</span>Age)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu. 
##    0.42   20.12   28.00   29.70   38.00 
##    Max.    NA&#39;s 
##   80.00     177</code></pre>
<p>Notice that there are 177 <code>NA</code> (missing) values for age.
Let’s visualize the missing data more systematically.</p>
</div>
<div id="explore-missing-data" class="section level2 unnumbered hasAnchor">
<h2>Explore missing data<a href="module-3-generalized-linear-models.html#explore-missing-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this let’s use a little script that converts a table into black and white squares to visualize missing data. For this, install the <code>plotrix</code> package</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb183-1"><a href="module-3-generalized-linear-models.html#cb183-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;plotrix&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;plotrix&quot;</span>)</span>
<span id="cb183-2"><a href="module-3-generalized-linear-models.html#cb183-2" tabindex="-1"></a></span>
<span id="cb183-3"><a href="module-3-generalized-linear-models.html#cb183-3" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">require</span>(plotrix))</span>
<span id="cb183-4"><a href="module-3-generalized-linear-models.html#cb183-4" tabindex="-1"></a></span>
<span id="cb183-5"><a href="module-3-generalized-linear-models.html#cb183-5" tabindex="-1"></a><span class="co">#&#39; show data missingness as a chequered matrix</span></span>
<span id="cb183-6"><a href="module-3-generalized-linear-models.html#cb183-6" tabindex="-1"></a><span class="co">#&#39; </span></span>
<span id="cb183-7"><a href="module-3-generalized-linear-models.html#cb183-7" tabindex="-1"></a><span class="co">#&#39; @param x (matrix) data matrix.</span></span>
<span id="cb183-8"><a href="module-3-generalized-linear-models.html#cb183-8" tabindex="-1"></a><span class="co">#&#39; @param outFile (char) path to file for printing graph</span></span>
<span id="cb183-9"><a href="module-3-generalized-linear-models.html#cb183-9" tabindex="-1"></a><span class="co">#&#39; @param wd (numeric) width in inches</span></span>
<span id="cb183-10"><a href="module-3-generalized-linear-models.html#cb183-10" tabindex="-1"></a><span class="co">#&#39; @param ht (numeric) height in inches</span></span>
<span id="cb183-11"><a href="module-3-generalized-linear-models.html#cb183-11" tabindex="-1"></a><span class="co">#&#39; @return plots missingness matrix to file</span></span>
<span id="cb183-12"><a href="module-3-generalized-linear-models.html#cb183-12" tabindex="-1"></a><span class="co">#&#39; @import plotrix</span></span>
<span id="cb183-13"><a href="module-3-generalized-linear-models.html#cb183-13" tabindex="-1"></a><span class="co">#&#39; @export</span></span>
<span id="cb183-14"><a href="module-3-generalized-linear-models.html#cb183-14" tabindex="-1"></a>plotMissMat <span class="ot">&lt;-</span> <span class="cf">function</span>(x,<span class="at">xlab=</span><span class="st">&quot;columns&quot;</span>,</span>
<span id="cb183-15"><a href="module-3-generalized-linear-models.html#cb183-15" tabindex="-1"></a>        <span class="at">ylab=</span><span class="st">&quot;rows&quot;</span>,<span class="at">border=</span><span class="cn">NA</span>) {</span>
<span id="cb183-16"><a href="module-3-generalized-linear-models.html#cb183-16" tabindex="-1"></a>    </span>
<span id="cb183-17"><a href="module-3-generalized-linear-models.html#cb183-17" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(x)</span>
<span id="cb183-18"><a href="module-3-generalized-linear-models.html#cb183-18" tabindex="-1"></a>    <span class="fu">class</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;numeric&quot;</span></span>
<span id="cb183-19"><a href="module-3-generalized-linear-models.html#cb183-19" tabindex="-1"></a>    <span class="fu">color2D.matplot</span>(x,<span class="at">show.values=</span><span class="cn">FALSE</span>,<span class="at">axes=</span><span class="cn">FALSE</span>,</span>
<span id="cb183-20"><a href="module-3-generalized-linear-models.html#cb183-20" tabindex="-1"></a>        <span class="at">cs1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs2=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">cs3=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>),<span class="at">border=</span>border,</span>
<span id="cb183-21"><a href="module-3-generalized-linear-models.html#cb183-21" tabindex="-1"></a>        <span class="at">cex=</span><span class="fl">0.8</span>,</span>
<span id="cb183-22"><a href="module-3-generalized-linear-models.html#cb183-22" tabindex="-1"></a>        <span class="at">xlab=</span>xlab,<span class="at">ylab=</span>ylab)</span>
<span id="cb183-23"><a href="module-3-generalized-linear-models.html#cb183-23" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s look at the missingness in the Titanic dataset. Missing data is shown as a white cell, and non-missing data is shown in black.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb184-1"><a href="module-3-generalized-linear-models.html#cb184-1" tabindex="-1"></a><span class="fu">plotMissMat</span>(dat)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-62-1.png" width="672" /></p>
<p>We can see a column with many missing values. This is probably the “age” data.
Let’s count the number of missing values on a per-column level.</p>
<p>For this we combine <code>is.na()</code>, which returns a TRUE/FALSE value for NA values, and <code>colSums()</code> which adds up the TRUE values down the columns.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb185-1"><a href="module-3-generalized-linear-models.html#cb185-1" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(dat))</span></code></pre></div>
<pre><code>## PassengerId    Survived      Pclass 
##           0           0           0 
##        Name         Sex         Age 
##           0           0         177 
##       SibSp       Parch      Ticket 
##           0           0           0 
##        Fare       Cabin    Embarked 
##           0           0           0</code></pre>
<p>This confirms that <code>Age</code> is the only column with missing data.</p>
<p>Now let’s explore the data using plots.</p>
</div>
<div id="essential-r-plots-with-ggplot2" class="section level2 unnumbered hasAnchor">
<h2>Essential R: Plots with <code>ggplot2</code><a href="module-3-generalized-linear-models.html#essential-r-plots-with-ggplot2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>ggplot2</code> is a popular plotting package that uses an additive paradigm to build plots.</p>
<p>Useful websites:
* <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">ggplot2 cheatsheet</a>
* The <a href="https://ggplot2.tidyverse.org/">ggplot2 website</a> is a wealth of reference information, so here we just touch on the basics to get you started.</p>
<p>Anytime you need to generate a specific kind of plot, the website will most likely have documentation for how to achieve it.</p>
<p>Let’s start by creating a scatterplot with two <strong>continuous variables</strong>. For this let’s load a dataset measuring statistics around quality of life in US states in the late 70’s:</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb187-1"><a href="module-3-generalized-linear-models.html#cb187-1" tabindex="-1"></a>state.x77 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(state.x77)</span>
<span id="cb187-2"><a href="module-3-generalized-linear-models.html#cb187-2" tabindex="-1"></a><span class="fu">head</span>(state.x77)</span></code></pre></div>
<pre><code>##            Population Income Illiteracy
## Alabama          3615   3624        2.1
## Alaska            365   6315        1.5
## Arizona          2212   4530        1.8
## Arkansas         2110   3378        1.9
## California      21198   5114        1.1
## Colorado         2541   4884        0.7
##            Life Exp Murder HS Grad Frost
## Alabama       69.05   15.1    41.3    20
## Alaska        69.31   11.3    66.7   152
## Arizona       70.55    7.8    58.1    15
## Arkansas      70.66   10.1    39.9    65
## California    71.71   10.3    62.6    20
## Colorado      72.06    6.8    63.9   166
##              Area
## Alabama     50708
## Alaska     566432
## Arizona    113417
## Arkansas    51945
## California 156361
## Colorado   103766</code></pre>
<p>Create a base plot using the <code>ggplot</code> function. Then “add” a scatterplot to it. Notice that the plot has been assigned to a variable named <code>p</code>.</p>
<p>This setup is standard for <code>ggplot2</code> and allows multiple visualizations to be applied to the same base plot.</p>
<p>We use <code>aes</code> to tell ggplot what the <code>x</code> and <code>y</code> axes are, and later, if we want to colour-code by a particular column.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb189-1"><a href="module-3-generalized-linear-models.html#cb189-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb189-2"><a href="module-3-generalized-linear-models.html#cb189-2" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(state.x77, </span>
<span id="cb189-3"><a href="module-3-generalized-linear-models.html#cb189-3" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> Illiteracy,<span class="at">y =</span> Income)</span>
<span id="cb189-4"><a href="module-3-generalized-linear-models.html#cb189-4" tabindex="-1"></a>    )</span>
<span id="cb189-5"><a href="module-3-generalized-linear-models.html#cb189-5" tabindex="-1"></a></span>
<span id="cb189-6"><a href="module-3-generalized-linear-models.html#cb189-6" tabindex="-1"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="co"># scatter plot</span></span>
<span id="cb189-7"><a href="module-3-generalized-linear-models.html#cb189-7" tabindex="-1"></a>p</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/load-states-data-1.png" width="672" /></p>
<p>Now let’s add confidence intervals:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb190-1"><a href="module-3-generalized-linear-models.html#cb190-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;
## and formula = &#39;y ~ x&#39;</code></pre>
<p><img src="bookdown-demo_files/figure-html/states-geomsmooth-1.png" width="672" /></p>
<p>It looks like there is a negative relationship between illiteracy and income.
We can confirm this by looking at correlation:</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb192-1"><a href="module-3-generalized-linear-models.html#cb192-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Illiteracy</span>
<span id="cb192-2"><a href="module-3-generalized-linear-models.html#cb192-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> state.x77<span class="sc">$</span>Income</span>
<span id="cb192-3"><a href="module-3-generalized-linear-models.html#cb192-3" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)</span></code></pre></div>
<pre><code>## 
##  Pearson&#39;s product-moment correlation
## 
## data:  x and y
## t = -3.3668, df = 48, p-value =
## 0.001505
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  -0.6378257 -0.1807128
## sample estimates:
##        cor 
## -0.4370752</code></pre>
<div class="sourceCode" id="cb194"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb194-1"><a href="module-3-generalized-linear-models.html#cb194-1" tabindex="-1"></a><span class="fu">cor.test</span>(x,y)<span class="sc">$</span>p.value</span></code></pre></div>
<pre><code>## [1] 0.001505073</code></pre>
<p>Let’s now examine <strong>categorical variables</strong>.
In the titanic set, let’s look at the fare paid based on passenger class:</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb196-1"><a href="module-3-generalized-linear-models.html#cb196-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat)</span>
<span id="cb196-2"><a href="module-3-generalized-linear-models.html#cb196-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_boxplot</span>(</span>
<span id="cb196-3"><a href="module-3-generalized-linear-models.html#cb196-3" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(Pclass), <span class="co"># &quot;factor()&quot; makes a data column a categorical variable</span></span>
<span id="cb196-4"><a href="module-3-generalized-linear-models.html#cb196-4" tabindex="-1"></a>        <span class="at">y =</span> Fare))</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/boxplot-1.png" width="672" /></p>
<p>We can use barplots to examine counts and proportions.
Let’s look at number of survivors, split by passenger class.</p>
<p>Here the <code>fill</code> command is used to to split each barplot by the category, “Survived”.
So you can see number of passengers by “Pclass” split by survival.</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb197-1"><a href="module-3-generalized-linear-models.html#cb197-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb197-2"><a href="module-3-generalized-linear-models.html#cb197-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), </span>
<span id="cb197-3"><a href="module-3-generalized-linear-models.html#cb197-3" tabindex="-1"></a>        Pclass)</span>
<span id="cb197-4"><a href="module-3-generalized-linear-models.html#cb197-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/barplot-1.png" width="672" /></p>
<p>The plot above shows count data. Let’s convert this to proportions. We can see that the fraction of non-survivors in “Class 3” is high.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb198-1"><a href="module-3-generalized-linear-models.html#cb198-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb198-2"><a href="module-3-generalized-linear-models.html#cb198-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), Pclass),</span>
<span id="cb198-3"><a href="module-3-generalized-linear-models.html#cb198-3" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb198-4"><a href="module-3-generalized-linear-models.html#cb198-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/barplot2%20-1.png" width="672" /></p>
<p>How about males versus females?</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb199-1"><a href="module-3-generalized-linear-models.html#cb199-1" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_bar</span>(</span>
<span id="cb199-2"><a href="module-3-generalized-linear-models.html#cb199-2" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">fill=</span><span class="fu">factor</span>(Survived), Sex),</span>
<span id="cb199-3"><a href="module-3-generalized-linear-models.html#cb199-3" tabindex="-1"></a>    <span class="at">position =</span> <span class="st">&quot;fill&quot;</span></span>
<span id="cb199-4"><a href="module-3-generalized-linear-models.html#cb199-4" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/finish-titanic-1.png" width="672" /></p>
<p><strong>Exercise for later:</strong> Try other <code>ggplot</code> functions on these data or on a small data table of your project.
(avoid using large genomics datasets, because those are going to be harder to interpret)</p>
</div>
<div id="fit-binary-response-variable-using-glm-and-logistic-regression" class="section level2 unnumbered hasAnchor">
<h2>Fit binary response variable using <code>glm()</code> and logistic regression<a href="module-3-generalized-linear-models.html#fit-binary-response-variable-using-glm-and-logistic-regression" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let’s fit a model to a binary outcome. For this we load a dataset that measures physiological variables in a cohort of Pima Indians.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb200-1"><a href="module-3-generalized-linear-models.html#cb200-1" tabindex="-1"></a><span class="fu">library</span>(mlbench)</span>
<span id="cb200-2"><a href="module-3-generalized-linear-models.html#cb200-2" tabindex="-1"></a><span class="fu">data</span>(PimaIndiansDiabetes2)</span>
<span id="cb200-3"><a href="module-3-generalized-linear-models.html#cb200-3" tabindex="-1"></a><span class="co"># type ?PimaIndiansDiabetes2 to learn more about the dataset.</span></span>
<span id="cb200-4"><a href="module-3-generalized-linear-models.html#cb200-4" tabindex="-1"></a></span>
<span id="cb200-5"><a href="module-3-generalized-linear-models.html#cb200-5" tabindex="-1"></a>dat <span class="ot">&lt;-</span> PimaIndiansDiabetes2</span>
<span id="cb200-6"><a href="module-3-generalized-linear-models.html#cb200-6" tabindex="-1"></a><span class="fu">head</span>(dat)</span></code></pre></div>
<pre><code>##   pregnant glucose pressure triceps
## 1        6     148       72      35
## 2        1      85       66      29
## 3        8     183       64      NA
## 4        1      89       66      23
## 5        0     137       40      35
## 6        5     116       74      NA
##   insulin mass pedigree age diabetes
## 1      NA 33.6    0.627  50      pos
## 2      NA 26.6    0.351  31      neg
## 3      NA 23.3    0.672  32      pos
## 4      94 28.1    0.167  21      neg
## 5     168 43.1    2.288  33      pos
## 6      NA 25.6    0.201  30      neg</code></pre>
<p>Let’s look at the impact of blood glucose levels on diabetes diagnosis.</p>
<p>First let’s make a scatterplot. Could there be a relationship?</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb202-1"><a href="module-3-generalized-linear-models.html#cb202-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(<span class="at">x =</span> glucose, <span class="at">y =</span> <span class="fu">factor</span>(diabetes)))</span>
<span id="cb202-2"><a href="module-3-generalized-linear-models.html#cb202-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<pre><code>## Warning: Removed 5 rows containing missing values
## or values outside the scale range
## (`geom_point()`).</code></pre>
<p><img src="bookdown-demo_files/figure-html/plot-diabetes-vs-glucose-1.png" width="672" /></p>
<p>Could this be fit with a linear model?</p>
<p>Is there a continuous line that could reasonably fit the data?</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb204-1"><a href="module-3-generalized-linear-models.html#cb204-1" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dat, <span class="fu">aes</span>(<span class="at">x =</span> glucose, <span class="at">y =</span> <span class="fu">factor</span>(diabetes)))</span>
<span id="cb204-2"><a href="module-3-generalized-linear-models.html#cb204-2" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>()</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;
## and formula = &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning: Removed 5 rows containing non-finite
## outside the scale range
## (`stat_smooth()`).</code></pre>
<pre><code>## Warning: Removed 5 rows containing missing values
## or values outside the scale range
## (`geom_point()`).</code></pre>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-64-1.png" width="672" /></p>
<p>This is a situation where a logistic regression model would be an appropriate choice because of the binary outcome.</p>
<p>We’re going to use <code>glm()</code> to fit a model to these data:</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb208-1"><a href="module-3-generalized-linear-models.html#cb208-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span>glucose, </span>
<span id="cb208-2"><a href="module-3-generalized-linear-models.html#cb208-2" tabindex="-1"></a>    dat,</span>
<span id="cb208-3"><a href="module-3-generalized-linear-models.html#cb208-3" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span> <span class="co"># set to model binary outcome</span></span>
<span id="cb208-4"><a href="module-3-generalized-linear-models.html#cb208-4" tabindex="-1"></a>)</span>
<span id="cb208-5"><a href="module-3-generalized-linear-models.html#cb208-5" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose, family = &quot;binomial&quot;, 
##     data = dat)
## 
## Coefficients:
##              Estimate Std. Error z value
## (Intercept) -5.715088   0.438100  -13.04
## glucose      0.040634   0.003382   12.01
##             Pr(&gt;|z|)    
## (Intercept)   &lt;2e-16 ***
## glucose       &lt;2e-16 ***
## ---
## Signif. codes:  
##   0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39;
##   0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 986.70  on 762  degrees of freedom
## Residual deviance: 786.56  on 761  degrees of freedom
##   (5 observations deleted due to missingness)
## AIC: 790.56
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>Which factors explain a diabetes diagnosis?</p>
<p>What if we include a couple other factors?</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb210-1"><a href="module-3-generalized-linear-models.html#cb210-1" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">factor</span>(diabetes)<span class="sc">~</span> glucose <span class="sc">+</span> pregnant <span class="sc">+</span> age <span class="sc">+</span> pressure <span class="sc">+</span> triceps,</span>
<span id="cb210-2"><a href="module-3-generalized-linear-models.html#cb210-2" tabindex="-1"></a>    dat,</span>
<span id="cb210-3"><a href="module-3-generalized-linear-models.html#cb210-3" tabindex="-1"></a>    <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb210-4"><a href="module-3-generalized-linear-models.html#cb210-4" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = factor(diabetes) ~ glucose + pregnant + age + pressure + 
##     triceps, family = &quot;binomial&quot;, data = dat)
## 
## Coefficients:
##               Estimate Std. Error z value
## (Intercept) -7.6685706  0.8451262  -9.074
## glucose      0.0365926  0.0041105   8.902
## pregnant     0.0972493  0.0417560   2.329
## age          0.0249707  0.0135945   1.837
## pressure    -0.0009321  0.0099322  -0.094
## triceps      0.0420045  0.0117044   3.589
##             Pr(&gt;|z|)    
## (Intercept)  &lt; 2e-16 ***
## glucose      &lt; 2e-16 ***
## pregnant    0.019860 *  
## age         0.066235 .  
## pressure    0.925230    
## triceps     0.000332 ***
## ---
## Signif. codes:  
##   0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39;
##   0.1 &#39; &#39; 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 678.40  on 533  degrees of freedom
## Residual deviance: 494.38  on 528  degrees of freedom
##   (234 observations deleted due to missingness)
## AIC: 506.38
## 
## Number of Fisher Scoring iterations: 5</code></pre>
<div class="rmd-caution">
<p>
Note: This morning’s session only intends to introduce you to fitting
non-linear models.
</p>
<p>
In practice you may need to do more work to test multiple models to
ascertain best fits your data, using measures such as goodness-of-fit.
You will also likely compute odds ratio (odds of increased Y per unit
increase X), which is out of scope for the current tutorial.
</p>
<p>
We strongly recommend that you learn these topics before applying
these methods to your own data.
</p>
</div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2-exercise-results.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4-finding-differentially-expressed-genes-from-rnaseq-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
